<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • numbat</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png"><meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="preparing-data">Preparing data<a class="anchor" aria-label="anchor" href="#preparing-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="aggregate_counts.html">aggregate_counts()</a></code>

        </dt>
        <dd>Utility function to make reference gene expression profiles</dd>
      </dl></div><div class="section level2">
      <h2 id="cnv-detection">CNV detection<a class="anchor" aria-label="anchor" href="#cnv-detection"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="get_bulk.html">get_bulk()</a></code>

        </dt>
        <dd>Aggregate single-cell data into combined bulk expression and allele profile</dd>
      </dl><dl><dt>

          <code><a href="analyze_bulk.html">analyze_bulk()</a></code>

        </dt>
        <dd>Call CNVs in a pseudobulk profile using the Numbat joint HMM</dd>
      </dl><dl><dt>

          <code><a href="detect_clonal_loh.html">detect_clonal_loh()</a></code>

        </dt>
        <dd>Call clonal LOH using SNP density. Rcommended for cell lines or tumor samples with no normal cells.</dd>
      </dl></div><div class="section level2">
      <h2 id="single-cell-clonal-decomposition">Single-cell clonal decomposition<a class="anchor" aria-label="anchor" href="#single-cell-clonal-decomposition"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="run_numbat.html">run_numbat()</a></code>

        </dt>
        <dd>Run workflow to decompose tumor subclones</dd>
      </dl></div><div class="section level2">
      <h2 id="visualizing-results">Visualizing results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="Numbat.html">Numbat</a></code>

        </dt>
        <dd>Numbat R6 class</dd>
      </dl><dl><dt>

          <code><a href="get_gtree.html">get_gtree()</a></code>

        </dt>
        <dd>Get a tidygraph tree with simplified mutational history.</dd>
      </dl><dl><dt>

          <code><a href="plot_bulks.html">plot_bulks()</a></code>

        </dt>
        <dd>Plot a group of pseudobulk HMM profiles</dd>
      </dl><dl><dt>

          <code><a href="plot_consensus.html">plot_consensus()</a></code>

        </dt>
        <dd>Plot consensus CNVs</dd>
      </dl><dl><dt>

          <code><a href="plot_exp_roll.html">plot_exp_roll()</a></code>

        </dt>
        <dd>Plot single-cell smoothed expression magnitude heatmap</dd>
      </dl><dl><dt>

          <code><a href="plot_mut_history.html">plot_mut_history()</a></code>

        </dt>
        <dd>Plot mutational history</dd>
      </dl><dl><dt>

          <code><a href="plot_phylo_heatmap.html">plot_phylo_heatmap()</a></code>

        </dt>
        <dd>Plot single-cell CNV calls along with the clonal phylogeny</dd>
      </dl><dl><dt>

          <code><a href="plot_psbulk.html">plot_psbulk()</a></code>

        </dt>
        <dd>Plot a pseudobulk HMM profile</dd>
      </dl><dl><dt>

          <code><a href="plot_sc_tree.html">plot_sc_tree()</a></code>

        </dt>
        <dd>Plot single-cell smoothed expression magnitude heatmap</dd>
      </dl><dl><dt>

          <code><a href="cnv_heatmap.html">cnv_heatmap()</a></code>

        </dt>
        <dd>Plot CNV heatmap</dd>
      </dl></div><div class="section level2">
      <h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="df_allele_example.html">df_allele_example</a></code>

        </dt>
        <dd>example allele count dataframe</dd>
      </dl><dl><dt>

          <code><a href="count_mat_example.html">count_mat_example</a></code>

        </dt>
        <dd>example gene expression count matrix</dd>
      </dl><dl><dt>

          <code><a href="ref_hca.html">ref_hca</a></code>

        </dt>
        <dd>reference expression magnitudes from HCA</dd>
      </dl><dl><dt>

          <code><a href="ref_hca_counts.html">ref_hca_counts</a></code>

        </dt>
        <dd>reference expression counts from HCA</dd>
      </dl><dl><dt>

          <code><a href="bulk_example.html">bulk_example</a></code>

        </dt>
        <dd>example pseudobulk dataframe</dd>
      </dl><dl><dt>

          <code><a href="annot_ref.html">annot_ref</a></code>

        </dt>
        <dd>example reference cell annotation</dd>
      </dl><dl><dt>

          <code><a href="count_mat_ref.html">count_mat_ref</a></code>

        </dt>
        <dd>example reference count matrix</dd>
      </dl><dl><dt>

          <code><a href="gexp_roll_example.html">gexp_roll_example</a></code>

        </dt>
        <dd>example smoothed gene expression dataframe</dd>
      </dl><dl><dt>

          <code><a href="hc_example.html">hc_example</a></code>

        </dt>
        <dd>example hclust tree</dd>
      </dl><dl><dt>

          <code><a href="joint_post_example.html">joint_post_example</a></code>

        </dt>
        <dd>example joint single-cell cnv posterior dataframe</dd>
      </dl><dl><dt>

          <code><a href="mut_graph_example.html">mut_graph_example</a></code>

        </dt>
        <dd>example mutation graph</dd>
      </dl><dl><dt>

          <code><a href="phylogeny_example.html">phylogeny_example</a></code>

        </dt>
        <dd>example single-cell phylogeny</dd>
      </dl><dl><dt>

          <code><a href="pre_likelihood_hmm.html">pre_likelihood_hmm</a></code>

        </dt>
        <dd>HMM object for unit tests</dd>
      </dl><dl><dt>

          <code><a href="segs_example.html">segs_example</a></code>

        </dt>
        <dd>example CNV segments dataframe</dd>
      </dl></div><div class="section level2">
      <h2 id="genome-annotations">Genome annotations<a class="anchor" aria-label="anchor" href="#genome-annotations"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="acen_hg19.html">acen_hg19</a></code>

        </dt>
        <dd>centromere regions (hg19)</dd>
      </dl><dl><dt>

          <code><a href="acen_hg38.html">acen_hg38</a></code>

        </dt>
        <dd>centromere regions (hg38)</dd>
      </dl><dl><dt>

          <code><a href="chrom_sizes_hg19.html">chrom_sizes_hg19</a></code>

        </dt>
        <dd>chromosome sizes (hg19)</dd>
      </dl><dl><dt>

          <code><a href="chrom_sizes_hg38.html">chrom_sizes_hg38</a></code>

        </dt>
        <dd>chromosome sizes (hg38)</dd>
      </dl><dl><dt>

          <code><a href="gaps_hg19.html">gaps_hg19</a></code>

        </dt>
        <dd>genome gap regions (hg19)</dd>
      </dl><dl><dt>

          <code><a href="gaps_hg38.html">gaps_hg38</a></code>

        </dt>
        <dd>genome gap regions (hg38)</dd>
      </dl><dl><dt>

          <code><a href="gtf_hg19.html">gtf_hg19</a></code>

        </dt>
        <dd>gene model (hg19)</dd>
      </dl><dl><dt>

          <code><a href="gtf_hg38.html">gtf_hg38</a></code>

        </dt>
        <dd>gene model (hg38)</dd>
      </dl><dl><dt>

          <code><a href="gtf_mm10.html">gtf_mm10</a></code>

        </dt>
        <dd>gene model (mm10)</dd>
      </dl><dl><dt>

          <code><a href="vcf_meta.html">vcf_meta</a></code>

        </dt>
        <dd>example VCF header</dd>
      </dl></div><div class="section level2">
      <h2 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="annotate_genes.html">annotate_genes()</a></code>

        </dt>
        <dd>Annotate genes on allele dataframe</dd>
      </dl><dl><dt>

          <code><a href="upgma.html">upgma()</a></code>

        </dt>
        <dd>UPGMA and WPGMA clustering</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

