<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Numbat R6 class — Numbat • numbat</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Numbat R6 class — Numbat"><meta name="description" content="Used to allow users to plot results"><meta property="og:description" content="Used to allow users to plot results"><meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png"><meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Numbat R6 class</h1>
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/main/R/class.R" class="external-link"><code>R/class.R</code></a></small>
      <div class="d-none name"><code>Numbat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Used to allow users to plot results</p>
    </div>


    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a new 'Numbat' object</p>
    </div>
    <div class="section level2">
    <h2 id="public-fields">Public fields<a class="anchor" aria-label="anchor" href="#public-fields"></a></h2>
    <p></p><div class="r6-fields"><dl><dt><code>label</code></dt>
<dd><p>character Sample name</p></dd>


<dt><code>gtf</code></dt>
<dd><p>dataframe Transcript annotation</p></dd>


<dt><code>joint_post</code></dt>
<dd><p>dataframe Joint posterior</p></dd>


<dt><code>exp_post</code></dt>
<dd><p>dataframe Expression posterior</p></dd>


<dt><code>allele_post</code></dt>
<dd><p>dataframe Allele posetrior</p></dd>


<dt><code>bulk_subtrees</code></dt>
<dd><p>dataframe Bulk profiles of lineage subtrees</p></dd>


<dt><code>bulk_clones</code></dt>
<dd><p>dataframe Bulk profiles of clones</p></dd>


<dt><code>segs_consensus</code></dt>
<dd><p>dataframe Consensus segments</p></dd>


<dt><code>tree_post</code></dt>
<dd><p>list Tree posterior</p></dd>


<dt><code>mut_graph</code></dt>
<dd><p>igraph Mutation history graph</p></dd>


<dt><code>gtree</code></dt>
<dd><p>tbl_graph Single-cell phylogeny</p></dd>


<dt><code>clone_post</code></dt>
<dd><p>dataframe Clone posteriors</p></dd>


<dt><code>gexp_roll_wide</code></dt>
<dd><p>matrix Smoothed expression of single cells</p></dd>


<dt><code>P</code></dt>
<dd><p>matrix Genotype probability matrix</p></dd>


<dt><code>treeML</code></dt>
<dd><p>matrix Maximum likelihood tree as phylo object</p></dd>


<dt><code>hc</code></dt>
<dd><p>hclust Initial hierarchical clustering</p></dd>


</dl><p></p></div>
    </div>
    <div class="section level2">
    <h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a></h2>

<div class="section">
<h3 id="public-methods">Public methods<a class="anchor" aria-label="anchor" href="#public-methods"></a></h3>

<ul><li><p><a href="#method-Numbat-new"><code>Numbat$new()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_phylo_heatmap"><code>Numbat$plot_phylo_heatmap()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_exp_roll"><code>Numbat$plot_exp_roll()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_mut_history"><code>Numbat$plot_mut_history()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_sc_tree"><code>Numbat$plot_sc_tree()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_consensus"><code>Numbat$plot_consensus()</code></a></p></li>
<li><p><a href="#method-Numbat-plot_clone_profile"><code>Numbat$plot_clone_profile()</code></a></p></li>
<li><p><a href="#method-Numbat-cutree"><code>Numbat$cutree()</code></a></p></li>
<li><p><a href="#method-Numbat-clone"><code>Numbat$clone()</code></a></p></li>
</ul></div><p></p><hr><a id="method-Numbat-new"></a><div class="section">
<h3 id="method-new-">Method <code>new()</code><a class="anchor" aria-label="anchor" href="#method-new-"></a></h3>
<p>initialize Numbat class</p><div class="section">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va"><a href="../reference/Numbat.html">Numbat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">out_dir</span>, i <span class="op">=</span> <span class="fl">2</span>, gtf <span class="op">=</span> <span class="va">gtf_hg38</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h4>
<p></p><div class="arguments"><dl><dt><code>out_dir</code></dt>
<dd><p>character string Output directory</p></dd>


<dt><code>i</code></dt>
<dd><p>integer Get results from which iteration (default=2)</p></dd>


<dt><code>gtf</code></dt>
<dd><p>dataframe Transcript gtf (default=gtf_hg38)</p></dd>


<dt><code>verbose</code></dt>
<dd><p>logical Whether to output verbose results (default=TRUE)</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns">Returns<a class="anchor" aria-label="anchor" href="#returns"></a></h4>
<p>a new 'Numbat' object</p>
</div>

</div><p></p><hr><a id="method-Numbat-plot_phylo_heatmap"></a><div class="section">
<h3 id="method-plot-phylo-heatmap-">Method <code><a href="plot_phylo_heatmap.html">plot_phylo_heatmap()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-phylo-heatmap-"></a></h3>
<p>Plot the single-cell CNV calls in a heatmap and the corresponding phylogeny</p><div class="section">
<h4 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-1">Arguments<a class="anchor" aria-label="anchor" href="#arguments-1"></a></h4>
<p></p><div class="arguments"><dl><dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_phylo_heatmap()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-plot_exp_roll"></a><div class="section">
<h3 id="method-plot-exp-roll-">Method <code><a href="plot_exp_roll.html">plot_exp_roll()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-exp-roll-"></a></h3>
<p>Plot window-smoothed expression profiles</p><div class="section">
<h4 id="usage-2">Usage<a class="anchor" aria-label="anchor" href="#usage-2"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_exp_roll</span><span class="op">(</span>k <span class="op">=</span> <span class="fl">3</span>, n_sample <span class="op">=</span> <span class="fl">300</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-2">Arguments<a class="anchor" aria-label="anchor" href="#arguments-2"></a></h4>
<p></p><div class="arguments"><dl><dt><code>k</code></dt>
<dd><p>integer Number of clusters</p></dd>


<dt><code>n_sample</code></dt>
<dd><p>integer Number of cells to subsample</p></dd>


<dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_exp_roll()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-plot_mut_history"></a><div class="section">
<h3 id="method-plot-mut-history-">Method <code><a href="plot_mut_history.html">plot_mut_history()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-mut-history-"></a></h3>
<p>Plot the mutation history of the tumor</p><div class="section">
<h4 id="usage-3">Usage<a class="anchor" aria-label="anchor" href="#usage-3"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_mut_history</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-3">Arguments<a class="anchor" aria-label="anchor" href="#arguments-3"></a></h4>
<p></p><div class="arguments"><dl><dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_mut_history()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-plot_sc_tree"></a><div class="section">
<h3 id="method-plot-sc-tree-">Method <code><a href="plot_sc_tree.html">plot_sc_tree()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-sc-tree-"></a></h3>
<p>Plot the single cell phylogeny</p><div class="section">
<h4 id="usage-4">Usage<a class="anchor" aria-label="anchor" href="#usage-4"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_sc_tree</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-4">Arguments<a class="anchor" aria-label="anchor" href="#arguments-4"></a></h4>
<p></p><div class="arguments"><dl><dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_sc_tree()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-plot_consensus"></a><div class="section">
<h3 id="method-plot-consensus-">Method <code><a href="plot_consensus.html">plot_consensus()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-consensus-"></a></h3>
<p>Plot consensus segments</p><div class="section">
<h4 id="usage-5">Usage<a class="anchor" aria-label="anchor" href="#usage-5"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_consensus</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-5">Arguments<a class="anchor" aria-label="anchor" href="#arguments-5"></a></h4>
<p></p><div class="arguments"><dl><dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_sc_tree()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-plot_clone_profile"></a><div class="section">
<h3 id="method-plot-clone-profile-">Method <code>plot_clone_profile()</code><a class="anchor" aria-label="anchor" href="#method-plot-clone-profile-"></a></h3>
<p>Plot clone cnv profiles</p><div class="section">
<h4 id="usage-6">Usage<a class="anchor" aria-label="anchor" href="#usage-6"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">plot_clone_profile</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-6">Arguments<a class="anchor" aria-label="anchor" href="#arguments-6"></a></h4>
<p></p><div class="arguments"><dl><dt><code>...</code></dt>
<dd><p>additional parameters passed to plot_clone_profile()</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-cutree"></a><div class="section">
<h3 id="method-cutree-">Method <code><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree()</a></code><a class="anchor" aria-label="anchor" href="#method-cutree-"></a></h3>
<p>Re-define subclones on the phylogeny.</p><div class="section">
<h4 id="usage-7">Usage<a class="anchor" aria-label="anchor" href="#usage-7"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">cutree</span><span class="op">(</span>max_cost <span class="op">=</span> <span class="fl">0</span>, n_cut <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-7">Arguments<a class="anchor" aria-label="anchor" href="#arguments-7"></a></h4>
<p></p><div class="arguments"><dl><dt><code>max_cost</code></dt>
<dd><p>numeric Likelihood threshold to collapse internal branches</p></dd>


<dt><code>n_cut</code></dt>
<dd><p>integer Number of cuts on the phylogeny to define subclones</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-Numbat-clone"></a><div class="section">
<h3 id="method-clone-">Method <code>clone()</code><a class="anchor" aria-label="anchor" href="#method-clone-"></a></h3>
<p>The objects of this class are cloneable with this method.</p><div class="section">
<h4 id="usage-8">Usage<a class="anchor" aria-label="anchor" href="#usage-8"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">Numbat</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span>deep <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-8">Arguments<a class="anchor" aria-label="anchor" href="#arguments-8"></a></h4>
<p></p><div class="arguments"><dl><dt><code>deep</code></dt>
<dd><p>Whether to make a deep clone.</p></dd>


</dl><p></p></div>
</div>

</div>

    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

