<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run workflow to decompose tumor subclones — run_numbat • numbat</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run workflow to decompose tumor subclones — run_numbat"><meta name="description" content="Run workflow to decompose tumor subclones"><meta property="og:description" content="Run workflow to decompose tumor subclones"><meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png"><meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Preparing Inputs for Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run workflow to decompose tumor subclones</h1>
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/v1.5.1/R/main.R" class="external-link"><code>R/main.R</code></a></small>
      <div class="d-none name"><code>run_numbat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run workflow to decompose tumor subclones</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_numbat</span><span class="op">(</span></span>
<span>  <span class="va">count_mat</span>,</span>
<span>  <span class="va">lambdas_ref</span>,</span>
<span>  <span class="va">df_allele</span>,</span>
<span>  gtf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  max_nni <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  t <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  gamma <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  min_LLR <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  max_entropy <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  init_k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min_cells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  tau <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  nu <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_cost <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">*</span> <span class="va">tau</span>,</span>
<span>  n_cut <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_depth <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  plot_min_depth <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  common_diploid <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_overlap <span class="op">=</span> <span class="fl">0.45</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ncores_nni <span class="op">=</span> <span class="va">ncores</span>,</span>
<span>  random_init <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  segs_loh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  call_clonal_loh <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  diploid_chroms <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  segs_consensus_fix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_loh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_genes <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  skip_nj <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  multi_allelic <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p_multi <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">alpha</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_convergence <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude_neu <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-count-mat">count_mat<a class="anchor" aria-label="anchor" href="#arg-count-mat"></a></dt>
<dd><p>dgCMatrix Raw count matrices where rownames are genes and column names are cells</p></dd>


<dt id="arg-lambdas-ref">lambdas_ref<a class="anchor" aria-label="anchor" href="#arg-lambdas-ref"></a></dt>
<dd><p>matrix Either a named vector with gene names as names and normalized expression as values, or a matrix where rownames are genes and columns are pseudobulk names</p></dd>


<dt id="arg-df-allele">df_allele<a class="anchor" aria-label="anchor" href="#arg-df-allele"></a></dt>
<dd><p>dataframe Allele counts per cell, produced by preprocess_allele</p></dd>


<dt id="arg-gtf">gtf<a class="anchor" aria-label="anchor" href="#arg-gtf"></a></dt>
<dd><p>dataframe Transcript GTF, if NULL will use the default GTF for the specified genome</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>character Genome version (hg38, hg19, or mm10)</p></dd>


<dt id="arg-out-dir">out_dir<a class="anchor" aria-label="anchor" href="#arg-out-dir"></a></dt>
<dd><p>string Output directory</p></dd>


<dt id="arg-max-iter">max_iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>integer Maximum number of iterations to run the phyologeny optimization</p></dd>


<dt id="arg-max-nni">max_nni<a class="anchor" aria-label="anchor" href="#arg-max-nni"></a></dt>
<dd><p>integer Maximum number of iterations to run NNI in the ML phylogeny inference</p></dd>


<dt id="arg-t">t<a class="anchor" aria-label="anchor" href="#arg-t"></a></dt>
<dd><p>numeric Transition probability</p></dd>


<dt id="arg-gamma">gamma<a class="anchor" aria-label="anchor" href="#arg-gamma"></a></dt>
<dd><p>numeric Dispersion parameter for the Beta-Binomial allele model</p></dd>


<dt id="arg-min-llr">min_LLR<a class="anchor" aria-label="anchor" href="#arg-min-llr"></a></dt>
<dd><p>numeric Minimum LLR to filter CNVs</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>numeric P value cutoff for diploid finding</p></dd>


<dt id="arg-eps">eps<a class="anchor" aria-label="anchor" href="#arg-eps"></a></dt>
<dd><p>numeric Convergence threshold for ML tree search</p></dd>


<dt id="arg-max-entropy">max_entropy<a class="anchor" aria-label="anchor" href="#arg-max-entropy"></a></dt>
<dd><p>numeric Entropy threshold to filter CNVs</p></dd>


<dt id="arg-init-k">init_k<a class="anchor" aria-label="anchor" href="#arg-init-k"></a></dt>
<dd><p>integer Number of clusters in the initial clustering</p></dd>


<dt id="arg-min-cells">min_cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>integer Minimum number of cells to run HMM on</p></dd>


<dt id="arg-tau">tau<a class="anchor" aria-label="anchor" href="#arg-tau"></a></dt>
<dd><p>numeric Factor to determine max_cost as a function of the number of cells (0-1)</p></dd>


<dt id="arg-nu">nu<a class="anchor" aria-label="anchor" href="#arg-nu"></a></dt>
<dd><p>numeric Phase switch rate</p></dd>


<dt id="arg-max-cost">max_cost<a class="anchor" aria-label="anchor" href="#arg-max-cost"></a></dt>
<dd><p>numeric Likelihood threshold to collapse internal branches</p></dd>


<dt id="arg-n-cut">n_cut<a class="anchor" aria-label="anchor" href="#arg-n-cut"></a></dt>
<dd><p>integer Number of cuts on the phylogeny to define subclones</p></dd>


<dt id="arg-min-depth">min_depth<a class="anchor" aria-label="anchor" href="#arg-min-depth"></a></dt>
<dd><p>integer Minimum allele depth</p></dd>


<dt id="arg-plot-min-depth">plot_min_depth<a class="anchor" aria-label="anchor" href="#arg-plot-min-depth"></a></dt>
<dd><p>integer Minimum allele depth to plot in pseudobulk HMM</p></dd>


<dt id="arg-common-diploid">common_diploid<a class="anchor" aria-label="anchor" href="#arg-common-diploid"></a></dt>
<dd><p>logical Whether to find common diploid regions in a group of peusdobulks</p></dd>


<dt id="arg-min-overlap">min_overlap<a class="anchor" aria-label="anchor" href="#arg-min-overlap"></a></dt>
<dd><p>numeric Minimum CNV overlap threshold</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>integer Number of threads to use</p></dd>


<dt id="arg-ncores-nni">ncores_nni<a class="anchor" aria-label="anchor" href="#arg-ncores-nni"></a></dt>
<dd><p>integer Number of threads to use for NNI</p></dd>


<dt id="arg-random-init">random_init<a class="anchor" aria-label="anchor" href="#arg-random-init"></a></dt>
<dd><p>logical Whether to initiate phylogney using a random tree (internal use only)</p></dd>


<dt id="arg-segs-loh">segs_loh<a class="anchor" aria-label="anchor" href="#arg-segs-loh"></a></dt>
<dd><p>dataframe Segments of clonal LOH to be excluded</p></dd>


<dt id="arg-call-clonal-loh">call_clonal_loh<a class="anchor" aria-label="anchor" href="#arg-call-clonal-loh"></a></dt>
<dd><p>logical Whether to call segments with clonal LOH</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical Verbosity</p></dd>


<dt id="arg-diploid-chroms">diploid_chroms<a class="anchor" aria-label="anchor" href="#arg-diploid-chroms"></a></dt>
<dd><p>vector Known diploid chromosomes</p></dd>


<dt id="arg-segs-consensus-fix">segs_consensus_fix<a class="anchor" aria-label="anchor" href="#arg-segs-consensus-fix"></a></dt>
<dd><p>dataframe Pre-determined segmentation of consensus CNVs</p></dd>


<dt id="arg-use-loh">use_loh<a class="anchor" aria-label="anchor" href="#arg-use-loh"></a></dt>
<dd><p>logical Whether to include LOH regions in the expression baseline</p></dd>


<dt id="arg-min-genes">min_genes<a class="anchor" aria-label="anchor" href="#arg-min-genes"></a></dt>
<dd><p>integer Minimum number of genes to call a segment</p></dd>


<dt id="arg-skip-nj">skip_nj<a class="anchor" aria-label="anchor" href="#arg-skip-nj"></a></dt>
<dd><p>logical Whether to skip NJ tree construction and only use UPGMA</p></dd>


<dt id="arg-multi-allelic">multi_allelic<a class="anchor" aria-label="anchor" href="#arg-multi-allelic"></a></dt>
<dd><p>logical Whether to call multi-allelic CNVs</p></dd>


<dt id="arg-p-multi">p_multi<a class="anchor" aria-label="anchor" href="#arg-p-multi"></a></dt>
<dd><p>numeric P value cutoff for calling multi-allelic CNVs</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>logical Whether to plot results</p></dd>


<dt id="arg-check-convergence">check_convergence<a class="anchor" aria-label="anchor" href="#arg-check-convergence"></a></dt>
<dd><p>logical Whether to terminate iterations based on consensus CNV convergence</p></dd>


<dt id="arg-exclude-neu">exclude_neu<a class="anchor" aria-label="anchor" href="#arg-exclude-neu"></a></dt>
<dd><p>logical Whether to exclude neutral segments from CNV retesting (internal use only)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a status code</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

