---
title: ""
documentclass: book
output:
  bookdown::gitbook:
    split_by: rmd
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

# Interpreting Numbat outputs
In this tutorial, we will illustrate how to visualize and interpret Numbat output, using a triple-negative breast cancer dataset (TNBC1) from [Gao et al](https://www.nature.com/articles/s41587-020-00795-2). We will use `pagoda2` for visualizing cells in low-dimensional expression space.

```{r}
library(ggplot2)
library(numbat)
library(dplyr)
library(glue)
library(data.table)
library(ggtree)
library(stringr)
library(tidygraph)
library(patchwork)
```

```{r}
# summarize output files into a Numbat object
nb = numbat$new(out_dir = '~/paper_data/numbat_out/TNBC1_demo')
# read saved pagoda2 object
pagoda = readRDS('~/paper_data/conos_objects/TNBC1.rds')
```

## Copy number landscape and single-cell phylogeny
As an overview, we can visualize the CNV calls in single-cells and their evolutionary relationships in an integrated view:
```{r, fig.width = 8, fig.height=3.5, dpi=300}
mypal = c('gray', "#377EB8", "#4DAF4A", "#984EA3")
  
nb$plot_phylo_heatmap(
  geno_bar = TRUE, 
  p_min = 0.9,
  pal_clone = mypal,
  multi_allelic = TRUE
)
```
In this visualization, the single-cell phylogeny (left) is juxtaposed with a heatmap of single-cell CNV calls (right). The CNV calls are colored by the type of alterations (AMP, amplification, BAMP, balanced amplification, DEL, deletion, CNLoH, copy-neutral loss of heterozygosity). The colorbar in-between differentiates the distinct genetic populations (genotype). The dashed blue line separates the predicted tumor versus normal cells. This tells us that the dataset mainly consists of three cell populations, a normal population (gray) and two tumor subclones (green and yellow). 

## Consensus copy number segments
Let's take a look at the consensus segments.
```{r, fig.width = 14, fig.height = 1.7, dpi=300}
nb$segs_consensus %>% plot_consensus()
```

## Bulk CNV profiles
We can also visualize these CNV events in pseudobulks where the data is more rich, aggregating cells by clone:
```{r, fig.width = 13, fig.height=6, dpi=300}

nb$bulk_clones %>% 
  plot_bulks(
    min_LLR = 50, # Log-likelihood ratio cutoff
    phi_mle = TRUE,
    legend = FALSE
  )
```

## Single-cell CNV calls
Numbat probabilistically evaluates the presence/absence of CNVs in single cells. The cell-level CNV posteriors are stored in the `nb$joint_post` dataframe:
```{r}
head(nb$joint_post) %>% select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y)
```
which contains cell-level information on specific CNV segments (`seg`), their alteration type (`cnv_state`), the joint posterior probability of the CNV (`p_cnv`), the expression-based posterior (`p_cnv_x`), and the allele-based posterior (`p_cnv_y`). We can visualize the event-specific posteriors in a expression-based tSNE embedding:
```{r, fig.width = 9, fig.height=3, dpi=300}

plist = list()

muts = c('1b', '3c', '4c')
cnv_type = nb$joint_post %>% distinct(seg, cnv_state) %>% {setNames(.$cnv_state, .$seg)}

for (mut in muts) {
    
    plist[[mut]] = pagoda$plotEmbedding(
        alpha=0.8,
        size=1, 
        plot.na = F, 
        colors = nb$joint_post %>%
            filter(seg == mut) %>%
            {setNames(.$p_cnv, .$cell)},
        show.legend = T,
        mark.groups = F,
        plot.theme = theme_bw(),
        title = paste0(mut, '(', cnv_type, ')')
    ) +
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1), name = 'Posterior')
}

wrap_plots(plist, guides = 'collect')
```

## Clonal assignments
Numbat aggregates signals across subclone-specific CNVs to probabilistically assign cells to subclones. The information regarding clonal assignments are contained in the `nb$clone_post` dataframe.
```{r}
nb$clone_post %>% head() %>% select(cell, clone_opt, p_1, p_2, p_3, p_4)
```
Here `clone_opt` denotes the maximum likelihood assignment of a cell to a specific clone. `p_{1..4}` are the detailed breakdown of the posterior probability that the cell belongs to each clone, respectively. Let's visualize the clonal decomposition in a tSNE embedding. Note that clone 1 is always the normal cells.
```{r, fig.width = 3, fig.height=3, dpi=300}
pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    groups = nb$clone_post %>%
        {setNames(.$clone_opt, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Genotypes',
    pal = mypal
)
```

## Tumor versus normal probability
Combining evidence from all CNVs, Numbat derives an aneuploidy probability for each cell to distinguish tumor versus normal cells. We can visualize the posterior aneuploidy probability based on expression evidence only, allele evidence only, and jointly:
```{r, fig.width = 9, fig.height=3, dpi=300}

p_joint = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Joint',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')

p_allele = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv_x, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Expression',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')

p_expr = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv_y, .$cell)},
    plot.na = F,
    show.legend = T,
    plot.theme = theme_bw(),
    title = 'Allele',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')

(p_expr | p_allele | p_joint) + plot_layout(guides = 'collect')

```
Both expression and allele signal clearly separate the tumor and normal cells.

## Tumor phylogeny
Let's take a closer look at the inferred single cell phylogeny and where the mutations occurred.
```{r, fig.width = 6, fig.height=3, dpi=300}
plot_sc_tree(
  nb$gtree, 
  label_size = 3, 
  branch_width = 0.5, 
  tip_length = 0.5,
  pal_clone = mypal, 
  tip = TRUE
)
```
The mutational history can also be represented on the clone level, where cells with the same genotype are aggregated into one node.
```{r, fig.width = 6, fig.height=3, dpi=300}
nb$plot_mut_history(pal_clone = mypal)
```



