<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>File descriptions • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="File descriptions">
<meta name="description" content="Full list of output files and column descriptions">
<meta property="og:description" content="Full list of output files and column descriptions">
<meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Preparing Inputs for Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>File descriptions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/HEAD/vignettes/descriptions.Rmd" class="external-link"><code>vignettes/descriptions.Rmd</code></a></small>
      <div class="d-none name"><code>descriptions.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="list-of-output-files">List of output files<a class="anchor" aria-label="anchor" href="#list-of-output-files"></a>
</h2>
<p>Numbat generates a number of files in the output folder. The file names are post-fixed with the <code>i</code>th iteration of phylogeny optimization. Here is a detailed list:</p>
<div class="section level5">
<h5 id="analysis-results">Analysis results<a class="anchor" aria-label="anchor" href="#analysis-results"></a>
</h5>
<ul>
<li>
<a href="#pseudobulk-profiles"><code>bulk_subtrees_{i}.tsv.gz</code></a>: Subtree pseudobulk profiles based on current cell lineage tree</li>
<li>
<a href="#consensus-segments"><code>segs_consensus_{i}.tsv.gz</code></a>: Consensus segments from subtree pseudobulk HMMs</li>
<li>
<a href="#pseudobulk-profiles"><code>bulk_clones_{i}.tsv.gz</code></a>: Clone-level pseudobulk profiles based on current cell lineage tree</li>
<li>
<a href="#exp_post"><code>exp_post_{i}.tsv</code></a>: Expression-based posterior probabilities of CNV states for each segment in each cell.</li>
<li>
<a href="#single-cell-posteriors"><code>allele_post_{i}.tsv</code></a>: Allele-based posterior probabilities of CNV states for each segment in each cell.</li>
<li>
<a href="#single-cell-posteriors"><code>joint_post_{i}.tsv</code></a>: Joint posterior probabilities of CNV states for each segment in each cell.</li>
<li>
<a href="#clone-assignments"><code>clone_post_{i}.tsv</code></a>: Single-cell clone assignment and tumor versus normal classification posteriors</li>
<li>
<a href="#pseudobulk-profiles"><code>bulk_clones_final.tsv.gz</code></a>: Clone-level pseudobulk profiles based on final cell lineage tree</li>
<li>
<code>bulk_subtrees_retest_{i}.tsv.gz</code>: Subtree pseudobulk profiles after retesting CNV states</li>
<li>
<code>gexp_roll_wide.tsv.gz</code>: window-smoothed normalized expression profiles of single cells</li>
<li>
<code>segs_loh.tsv</code>: Clonal LoH segments; written if <code>call_clonal_loh</code> is enabled</li>
</ul>
</div>
<div class="section level5">
<h5 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h5>
<ul>
<li>
<code>exp_roll_clust.png</code>: visualization of single-cell smoothed gene expression profiles</li>
<li>
<code>bulk_subtrees_{i}.png</code>: visualization of subtree pseudobulk CNV profiles</li>
<li>
<code>bulk_clones_{i}.png</code>: visualization of clone pseudobulk CNV profiles</li>
<li>
<code>bulk_clones_final.png</code>: visualization of final clone pseudobulk CNV profiles</li>
<li>
<code>tree_list_{i}.rds</code>: list of candidate phylogeneies in the maximum likelihood tree search</li>
<li>
<code>panel_{i}.png</code>: integrated visualization of single-cell phylogeny and CNV landscape</li>
</ul>
</div>
<div class="section level5">
<h5 id="lineage-trees">Lineage trees<a class="anchor" aria-label="anchor" href="#lineage-trees"></a>
</h5>
<ul>
<li>
<code>hc.rds</code>: initial hierarchical clustering result based on smoothed expression</li>
<li>
<code>clones_{i}.rds</code>: list of candidate clones used to generate <code>bulk_clones_{i}.tsv.gz</code>
</li>
<li>
<code>subtrees_{i}.rds</code>: list of candidate subtrees used to generate <code>bulk_subtrees_{i}.tsv.gz</code>
</li>
<li>
<code>tree_NJ_{i}.rds</code>: neighbor joining tree</li>
<li>
<code>mut_graph_{i}.rds</code>: mutation graph derived from the current cell lineage tree</li>
<li>
<code>tree_ML_{i}.rds</code>: maximum likelihood tree (in ape::phylo format)</li>
<li>
<code>tree_final_{i}.rds</code>: final cell lineage tree with mutation and clone annotation (in tbl_graph format)</li>
</ul>
</div>
<div class="section level5">
<h5 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a>
</h5>
<ul>
<li>
<code>log.txt</code>: log file</li>
<li>
<code>sc_refs.rds</code>: list of best reference match for each cell</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="single-cell-posteriors">Single-cell posteriors<a class="anchor" aria-label="anchor" href="#single-cell-posteriors"></a>
</h2>
<p><code>cell</code>: character; Cell barcode</p>
<p><code>CHROM</code>: character; Chromosome</p>
<p><code>seg</code>: character; Segment ID</p>
<p><code>cnv_state</code>: character; CNV state estimated from pseudobulk HMM</p>
<p><code>n_snp</code>: numeric; Number of SNPs in segment</p>
<p><code>seg_start</code>: numeric; Segment start position</p>
<p><code>seg_end</code>: numeric; Segment end position</p>
<p><code>n_genes</code>: numeric; Number of genes in segment</p>
<p><code>n_snps</code>: numeric; Number of SNPs in segment</p>
<p><code>prior_loh</code>: numeric; Prior probability of CNLoH</p>
<p><code>prior_amp</code>: numeric; Prior probability of amplification</p>
<p><code>prior_del</code>: numeric; Prior probability of deletion</p>
<p><code>prior_bamp</code>: numeric; Prior probability of biallelic amplification</p>
<p><code>prior_bdel</code>: numeric; Prior probability of biallelic deletion</p>
<p><code>l11_x</code>: numeric; Log-likelihood of CNV state 1:1 (neutral) given expression data</p>
<p><code>l20_x</code>: numeric; Log-likelihood of CNV state 2:0 (CNLoH) given expression data</p>
<p><code>l10_x</code>: numeric; Log-likelihood of CNV state 1:0 (deletion) given expression data</p>
<p><code>l21_x</code>: numeric; Log-likelihood of CNV state 2:1 (amplification) given expression data</p>
<p><code>l31_x</code>: numeric; Log-likelihood of CNV state 3:1 (amplification) given expression data</p>
<p><code>l22_x</code>: numeric; Log-likelihood of CNV state 2:2 (biallelic amplification) given expression data</p>
<p><code>l00_x</code>: numeric; Log-likelihood of CNV state 0:0 (biallelic deletion) given expression data</p>
<p><code>Z_cnv_x</code>: numeric; Total log(likelihood * prior) of CNV states given expression data</p>
<p><code>Z_n_x</code>: numeric; Total log(likelihood * prior) of neutral state given expression data</p>
<p><code>logBF_x</code>: numeric; Log Bayes factor of CNV state vs. neutral state given expression data</p>
<p><code>l11_y</code>: numeric; Log-likelihood of CNV state 1:1 (neutral) given allele data</p>
<p><code>l20_y</code>: numeric; Log-likelihood of CNV state 2:0 (CNLoH) given allele data</p>
<p><code>l10_y</code>: numeric; Log-likelihood of CNV state 1:0 (deletion) given allele data</p>
<p><code>l21_y</code>: numeric; Log-likelihood of CNV state 2:1 (gain) given allele data</p>
<p><code>l31_y</code>: numeric; Log-likelihood of CNV state 3:1 (amplification) given allele data</p>
<p><code>l22_y</code>: numeric; Log-likelihood of CNV state 2:2 (biallelic amplification) given allele data</p>
<p><code>l00_y</code>: numeric; Log-likelihood of CNV state 0:0 (biallelic deletion) given allele data</p>
<p><code>Z_cnv_y</code>: numeric; Total log(likelihood * prior) of CNV states given allele data</p>
<p><code>Z_n_y</code>: numeric; Total log(likelihood * prior) of neutral state given allele data</p>
<p><code>logBF_y</code>: numeric; Log Bayes factor of CNV state vs. neutral state given allele data</p>
<p><code>LLR</code>: numeric; Log-likelihood ratio of CNV state vs. neutral state</p>
<p><code>LLR_x</code>: numeric; Log-likelihood ratio of CNV state vs. neutral state given expression data</p>
<p><code>LLR_y</code>: numeric; Log-likelihood ratio of CNV state vs. neutral state given allele data</p>
<p><code>l11</code>: numeric; Joint log-likelihood of CNV state 1:1 (neutral)</p>
<p><code>l20</code>: numeric; Joint log-likelihood of CNV state 2:0 (CNLoH)</p>
<p><code>l10</code>: numeric; Joint log-likelihood of CNV state 1:0 (deletion)</p>
<p><code>l21</code>: numeric; Joint log-likelihood of CNV state 2:1 (amplification)</p>
<p><code>l31</code>: numeric; Joint log-likelihood of CNV state 3:1 (amplification)</p>
<p><code>l22</code>: numeric; Joint log-likelihood of CNV state 2:2 (biallelic amplification)</p>
<p><code>l00</code>: numeric; Joint log-likelihood of CNV state 0:0 (biallelic deletion)</p>
<p><code>Z_amp</code>: numeric; Total log(likelihood * prior) of amplification state</p>
<p><code>Z_loh</code>: numeric; Total log(likelihood * prior) of CNLoH state</p>
<p><code>Z_del</code>: numeric; Total log(likelihood * prior) of deletion state</p>
<p><code>Z_bamp</code>: numeric; Total log(likelihood * prior) of biallelic amplification state</p>
<p><code>Z_bdel</code>: numeric; Total log(likelihood * prior) of biallelic deletion state</p>
<p><code>Z_n</code>: numeric; Total log(likelihood * prior) of neutral state</p>
<p><code>Z</code>: numeric; Total log(likelihood * prior) of all states</p>
<p><code>Z_cnv</code>: numeric; Total log(likelihood * prior) of CNV states</p>
<p><code>p_amp</code>: numeric; Joint posterior probability of amplification states (2:1, 3:1)</p>
<p><code>p_neu</code>: numeric; Joint posterior probability of neutral state</p>
<p><code>p_del</code>: numeric; Joint posterior probability of deletion state</p>
<p><code>p_loh</code>: numeric; Joint posterior probability of CNLoH state</p>
<p><code>p_bamp</code>: numeric; Joint posterior probability of biallelic amplification state</p>
<p><code>p_bdel</code>: numeric; Joint posterior probability of biallelic deletion state</p>
<p><code>logBF</code>: numeric; Joint log Bayes factor of CNV state vs. neutral state</p>
<p><code>p_cnv</code>: numeric; Joint posterior probability of CNV state</p>
<p><code>p_n</code>: numeric; Joint posterior probability of neutral state</p>
<p><code>p_cnv_x</code>: numeric; Joint posterior probability of CNV state given expression data</p>
<p><code>p_cnv_y</code>: numeric; Joint posterior probability of CNV state given allele data</p>
<p><code>cnv_state_mle</code>: character; Maximum likelihood CNV state</p>
<p><code>cnv_state_map</code>: character; Maximum a posteriori CNV state</p>
<p><code>seg_label</code>: character; Segment label</p>
<p><code>avg_entropy</code>: numeric; Average entropy of CNV posterior in single cells</p>
<p><code>phi_mle</code>: numeric; Maximum likelihood of total copy number ratio relative to diploid (phi)</p>
<p><code>mu</code>: numeric; Mean of expression count distribution in the cell (Poisson log-Normal)</p>
<p><code>sigma</code>: numeric; Standard deviation of expression count distribution in the cell (Poisson log-Normal)</p>
<p><code>ref</code>: character; Best-matching single-cell expression reference</p>
<p><code>major</code>: integer; Major allele count; total SNP pileup counts deriving from the major haplotype in the CNV region for the given cell</p>
<p><code>minor</code>: integer; Minor allele count; total SNP pileup counts deriving from the minor haplotype in the CNV region for the given cell</p>
<p><code>total</code>: integer; Total allele count</p>
<p><code>MAF</code>: numeric; Major allele frequency</p>
</div>
<div class="section level2">
<h2 id="pseudobulk-profiles">Pseudobulk profiles<a class="anchor" aria-label="anchor" href="#pseudobulk-profiles"></a>
</h2>
<p><code>snp_id</code>: character; SNP ID</p>
<p><code>CHROM</code>: character; Chromosome</p>
<p><code>POS</code>: integer; Genomic position</p>
<p><code>cM</code>: numeric; Genetic distance in cM</p>
<p><code>REF</code>: character; Reference allele</p>
<p><code>ALT</code>: character; Alternate allele</p>
<p><code>GT</code>: character; Phased genotype</p>
<p><code>gene</code>: character; Gene symbol</p>
<p><code>AD</code>: integer; Allelic depth</p>
<p><code>DP</code>: integer; Total depth</p>
<p><code>AR</code>: numeric; Allelic fraction</p>
<p><code>snp_index</code>: integer; SNP index</p>
<p><code>pBAF</code>: numeric; Phased BAF</p>
<p><code>pAD</code>: numeric; Phased allelic depth</p>
<p><code>inter_snp_cm</code>: numeric; Genetic distance in cM between adjacent SNPs</p>
<p><code>p_s</code>: numeric; Probability of phase switch based on inter-SNP distance</p>
<p><code>Y_obs</code>: integer; Observed gene expression counts (the X/Y notation is switched here..)</p>
<p><code>lambda_obs</code>: numeric; Observed gene expression magnitude</p>
<p><code>lambda_ref</code>: numeric; Reference gene expression magnitude</p>
<p><code>d_obs</code>: numeric; Total gene expression count in the cell</p>
<p><code>gene_start</code>: integer; Gene start position</p>
<p><code>gene_end</code>: integer; Gene end position</p>
<p><code>gene_length</code>: integer; Gene length</p>
<p><code>gene_index</code>: integer; Gene index</p>
<p><code>logFC</code>: numeric; Log2 fold change of gene expression</p>
<p><code>lnFC</code>: numeric; Natural log fold change of gene expression</p>
<p><code>mse</code>: numeric; Mean squared error of logFC</p>
<p><code>snp_rate</code>: numeric; SNP rate</p>
<p><code>loh</code>: logical; True if segment has a clonal LoH (deletion)</p>
<p><code>n_cells</code>: integer; Number of cells in the pseudobulk</p>
<p><code>members</code>: character; Cell groups included in the pseudobulk</p>
<p><code>sample</code>: character/integer; Sample ID</p>
<p><code>state</code>: character; CNV state</p>
<p><code>boundary</code>: logical; True if the marker is at CNV boundary</p>
<p><code>seg_start_index</code>: integer; Segment start marker index</p>
<p><code>seg_end_index</code>: integer; Segment end marker index</p>
<p><code>seg_start</code>: integer; Segment start position</p>
<p><code>seg_end</code>: integer; Segment end position</p>
<p><code>seg_length</code>: integer; Segment length in bp</p>
<p><code>seg</code>: character; Segment ID</p>
<p><code>seg_cons</code>: character; Consensus segment ID</p>
<p><code>diploid</code>: logical; True if the segment is diploid</p>
<p><code>mu</code>: numeric; Mean of expression count distribution in the pseudobulk (Poisson log-Normal)</p>
<p><code>sig</code>: numeric; Standard deviation of expression count distribution in the pseudobulk (Poisson log-Normal)</p>
<p><code>cnv_state</code>: character; CNV state</p>
<p><code>n_genes</code>: integer; Number of genes in the segment</p>
<p><code>n_snps</code>: integer; Number of SNPs in the segment</p>
<p><code>theta_hat</code>: numeric; Crude estimate haplotype frequency based on MAF</p>
<p><code>theta_mle</code>: numeric; Maximum likelihood estimate of haplotype frequency (theta)</p>
<p><code>theta_sigma</code>: numeric; Standard deviation of theta MLE</p>
<p><code>L_y_n</code>: numeric; Neutral log-likelihood of allele data</p>
<p><code>L_y_d</code>: numeric; Deletion log-likelihood of allele data</p>
<p><code>L_y_a</code>: numeric; Amplification log-likelihood of allele data</p>
<p><code>phi_mle</code>: numeric; Maximum likelihood of total copy number ratio relative to diploid (phi)</p>
<p><code>phi_sigma</code>: numeric; Standard deviation of phi MLE</p>
<p><code>L_x_n</code>: numeric; Neutral log-likelihood of expression data</p>
<p><code>L_x_d</code>: numeric; Deletion log-likelihood of expression data</p>
<p><code>L_x_a</code>: numeric; Amplification log-likelihood of expression data</p>
<p><code>Z_cnv</code>: numeric; Total log(likelihood * prior) of CNV states</p>
<p><code>Z_n</code>: numeric; Total log(likelihood * prior) of neutral state</p>
<p><code>Z</code>: numeric; Total log(likelihood * prior) of all states</p>
<p><code>logBF</code>: numeric; Joint log Bayes factor of CNV state vs. neutral state</p>
<p><code>p_neu</code>: numeric; Joint posterior probability of neutral state</p>
<p><code>p_loh</code>: numeric; Joint posterior probability of CNLoH state</p>
<p><code>p_amp</code>: numeric; Joint posterior probability of amplification states (2:1, 3:1)</p>
<p><code>p_del</code>: numeric; Joint posterior probability of deletion state</p>
<p><code>p_bamp</code>: numeric; Joint posterior probability of biallelic amplification state</p>
<p><code>p_bdel</code>: numeric; Joint posterior probability of biallelic deletion state</p>
<p><code>LLR_x</code>: numeric; Log-likelihood ratio of expression data</p>
<p><code>LLR_y</code>: numeric; Log-likelihood ratio of allele data</p>
<p><code>LLR</code>: numeric; Log-likelihood ratio of all data</p>
<p><code>cnv_state_post</code>: character; Maximum a posteriori (retest) CNV state</p>
<p><code>state_post</code>: character; Maximum a posteriori CNV allelic state</p>
<p><code>p_up</code>: numeric; HMM posterior probability of variant allele belonging to the major haplotype</p>
<p><code>haplo_post</code>: numeric; Maximum a posteriori haplotype state assignment</p>
<p><code>haplo_naive</code>: numeric; Naive haplotype state assignment (based on BAF)</p>
<p><code>major_count</code>: integer; Major allele count; total SNP counts deriving from the major haplotype in the CNV region in the pseudobulk</p>
<p><code>minor_count</code>: integer; Minor allele count; total SNP counts deriving from the minor haplotype in the CNV region in the pseudobulk</p>
<p><code>theta_hat_roll</code>: numeric; Crude estimate haplotype frequency based on MAF (rolling window)</p>
<p><code>phi_mle_roll</code>: numeric; Maximum likelihood of total copy number ratio relative to diploid (phi, rolling window)</p>
<p><code>nu</code>: numeric; Phase-switch rate used in the HMM</p>
<p><code>gamma</code>: numeric; Allele inverse-overdispersion used in the HMM</p>
</div>
<div class="section level2">
<h2 id="consensus-segments">Consensus segments<a class="anchor" aria-label="anchor" href="#consensus-segments"></a>
</h2>
<p>consensus segments from subtree pseudobulk HMMs</p>
<p><code>sample</code>: character/integer; Sample ID</p>
<p><code>CHROM</code>: character; Chromosome</p>
<p><code>seg</code>: character; Segment ID</p>
<p><code>cnv_state</code>: character; CNV state</p>
<p><code>cnv_state_post</code>: character; Maximum a posteriori (retest) CNV state</p>
<p><code>seg_start</code>: integer; Segment start position</p>
<p><code>seg_end</code>: integer; Segment end position</p>
<p><code>seg_start_index</code>: integer; Segment start marker index</p>
<p><code>seg_end_index</code>: integer; Segment end marker index</p>
<p><code>theta_mle</code>: numeric; Maximum likelihood estimate of haplotype frequency (theta)</p>
<p><code>theta_sigma</code>: numeric; Standard deviation of theta MLE</p>
<p><code>phi_mle</code>: numeric; Maximum likelihood of total copy number ratio relative to diploid (phi)</p>
<p><code>phi_sigma</code>: numeric; Standard deviation of phi MLE</p>
<p><code>p_loh</code>: numeric; Joint posterior probability of CNLoH state</p>
<p><code>p_del</code>: numeric; Joint posterior probability of deletion state</p>
<p><code>p_amp</code>: numeric; Joint posterior probability of amplification states (2:1, 3:1)</p>
<p><code>p_bamp</code>: numeric; Joint posterior probability of biallelic amplification state</p>
<p><code>p_bdel</code>: numeric; Joint posterior probability of biallelic deletion state</p>
<p><code>LLR</code>: numeric; Log-likelihood ratio of all data</p>
<p><code>LLR_y</code>: numeric; Log-likelihood ratio of allele data</p>
<p><code>LLR_x</code>: numeric; Log-likelihood ratio of expression data</p>
<p><code>n_genes</code>: integer; Number of genes in the segment</p>
<p><code>n_snps</code>: integer; Number of SNPs in the segment</p>
<p><code>component</code>: integer; Component ID</p>
<p><code>LLR_sample</code>: numeric; Log-likelihood ratio in the sample where the CNV has the highest LLR</p>
<p><code>seg_length</code>: integer; Segment length in bp</p>
<p><code>seg_cons</code>: character; Consensus segment ID</p>
<p><code>n_states</code>: integer; Number of CNV states</p>
<p><code>cnv_states</code>: character; CNV states</p>
</div>
<div class="section level2">
<h2 id="clone-assignments">Clone assignments<a class="anchor" aria-label="anchor" href="#clone-assignments"></a>
</h2>
<p><code>cell</code>: character; Cell ID</p>
<p><code>clone_opt</code>: integer; Maximum a posteriori clone assignment</p>
<p><code>GT_opt</code>: character; Maximum a posteriori genotype</p>
<p><code>p_opt</code>: numeric; Maximum a posteriori clone probability</p>
<p><code>p_{k}</code>: numeric; Posterior probability of cell beloning to clone k</p>
<p><code>p_x_{k}</code>: numeric; Posterior probability of cell belonging to clone k given expression data</p>
<p><code>p_y_{k}</code>: numeric; Posterior probability of cell belonging to clone k given allele data</p>
<p><code>p_cnv</code>: numeric; Posterior probability of cell belonging to an aneuploid clone</p>
<p><code>p_cnv_x</code>: numeric; Posterior probability of cell belonging to an aneuploid clone given expression data</p>
<p><code>p_cnv_y</code>: numeric; Posterior probability of cell belonging to an aneuploid clone given allele data</p>
<p><code>compartment_opt</code>: character; Maximum a posteriori compartment (tumor vs normal) assignment</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
