<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot single-cell CNV calls along with the clonal phylogeny — plot_phylo_heatmap • numbat</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot single-cell CNV calls along with the clonal phylogeny — plot_phylo_heatmap"><meta name="description" content="Plot single-cell CNV calls along with the clonal phylogeny"><meta property="og:description" content="Plot single-cell CNV calls along with the clonal phylogeny"><meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png"><meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Preparing Inputs for Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plot single-cell CNV calls along with the clonal phylogeny</h1>
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/HEAD/R/vis.R" class="external-link"><code>R/vis.R</code></a></small>
      <div class="d-none name"><code>plot_phylo_heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot single-cell CNV calls along with the clonal phylogeny</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">gtree</span>,</span>
<span>  <span class="va">joint_post</span>,</span>
<span>  <span class="va">segs_consensus</span>,</span>
<span>  clone_post <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p_min <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  annot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pal_annot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot_title <span class="op">=</span> <span class="st">"Annotation"</span>,</span>
<span>  annot_scale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clone_dict <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clone_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clone_stack <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pal_clone <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clone_title <span class="op">=</span> <span class="st">"Genotype"</span>,</span>
<span>  clone_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  line_width <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  tree_height <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  branch_width <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  tip_length <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  annot_bar_width <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  clone_bar_width <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  bar_label_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  tvn_line <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clone_line <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude_gap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  root_edge <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  raster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_phylo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gtree">gtree<a class="anchor" aria-label="anchor" href="#arg-gtree"></a></dt>
<dd><p>tbl_graph The single-cell phylogeny</p></dd>


<dt id="arg-joint-post">joint_post<a class="anchor" aria-label="anchor" href="#arg-joint-post"></a></dt>
<dd><p>dataframe Joint single cell CNV posteriors</p></dd>


<dt id="arg-segs-consensus">segs_consensus<a class="anchor" aria-label="anchor" href="#arg-segs-consensus"></a></dt>
<dd><p>datatframe Consensus segment dataframe</p></dd>


<dt id="arg-clone-post">clone_post<a class="anchor" aria-label="anchor" href="#arg-clone-post"></a></dt>
<dd><p>dataframe Clone assignment posteriors</p></dd>


<dt id="arg-p-min">p_min<a class="anchor" aria-label="anchor" href="#arg-p-min"></a></dt>
<dd><p>numeric Probability threshold to display CNV calls</p></dd>


<dt id="arg-annot">annot<a class="anchor" aria-label="anchor" href="#arg-annot"></a></dt>
<dd><p>dataframe Cell annotations, dataframe with 'cell' and additional annotation columns</p></dd>


<dt id="arg-pal-annot">pal_annot<a class="anchor" aria-label="anchor" href="#arg-pal-annot"></a></dt>
<dd><p>named vector Colors for cell annotations</p></dd>


<dt id="arg-annot-title">annot_title<a class="anchor" aria-label="anchor" href="#arg-annot-title"></a></dt>
<dd><p>character Legend title for the annotation bar</p></dd>


<dt id="arg-annot-scale">annot_scale<a class="anchor" aria-label="anchor" href="#arg-annot-scale"></a></dt>
<dd><p>ggplot scale Color scale for the annotation bar</p></dd>


<dt id="arg-clone-dict">clone_dict<a class="anchor" aria-label="anchor" href="#arg-clone-dict"></a></dt>
<dd><p>named vector Clone annotations, mapping from cell name to clones</p></dd>


<dt id="arg-clone-bar">clone_bar<a class="anchor" aria-label="anchor" href="#arg-clone-bar"></a></dt>
<dd><p>logical Whether to display clone bar plot</p></dd>


<dt id="arg-clone-stack">clone_stack<a class="anchor" aria-label="anchor" href="#arg-clone-stack"></a></dt>
<dd><p>character Whether to plot clone assignment probabilities as stacked bar</p></dd>


<dt id="arg-pal-clone">pal_clone<a class="anchor" aria-label="anchor" href="#arg-pal-clone"></a></dt>
<dd><p>named vector Clone colors</p></dd>


<dt id="arg-clone-title">clone_title<a class="anchor" aria-label="anchor" href="#arg-clone-title"></a></dt>
<dd><p>character Legend title for the clone bar</p></dd>


<dt id="arg-clone-legend">clone_legend<a class="anchor" aria-label="anchor" href="#arg-clone-legend"></a></dt>
<dd><p>logical Whether to display the clone legend</p></dd>


<dt id="arg-line-width">line_width<a class="anchor" aria-label="anchor" href="#arg-line-width"></a></dt>
<dd><p>numeric Line width for CNV heatmap</p></dd>


<dt id="arg-tree-height">tree_height<a class="anchor" aria-label="anchor" href="#arg-tree-height"></a></dt>
<dd><p>numeric Relative height of the phylogeny plot</p></dd>


<dt id="arg-branch-width">branch_width<a class="anchor" aria-label="anchor" href="#arg-branch-width"></a></dt>
<dd><p>numeric Line width in the phylogeny</p></dd>


<dt id="arg-tip-length">tip_length<a class="anchor" aria-label="anchor" href="#arg-tip-length"></a></dt>
<dd><p>numeric Length of tips in the phylogeny</p></dd>


<dt id="arg-annot-bar-width">annot_bar_width<a class="anchor" aria-label="anchor" href="#arg-annot-bar-width"></a></dt>
<dd><p>numeric Width of annotation bar</p></dd>


<dt id="arg-clone-bar-width">clone_bar_width<a class="anchor" aria-label="anchor" href="#arg-clone-bar-width"></a></dt>
<dd><p>numeric Width of clone genotype bar</p></dd>


<dt id="arg-bar-label-size">bar_label_size<a class="anchor" aria-label="anchor" href="#arg-bar-label-size"></a></dt>
<dd><p>numeric Size of sidebar text labels</p></dd>


<dt id="arg-tvn-line">tvn_line<a class="anchor" aria-label="anchor" href="#arg-tvn-line"></a></dt>
<dd><p>logical Whether to draw line separating tumor and normal cells</p></dd>


<dt id="arg-clone-line">clone_line<a class="anchor" aria-label="anchor" href="#arg-clone-line"></a></dt>
<dd><p>logical Whether to display borders for clones in the heatmap</p></dd>


<dt id="arg-exclude-gap">exclude_gap<a class="anchor" aria-label="anchor" href="#arg-exclude-gap"></a></dt>
<dd><p>logical Whether to mark gap regions</p></dd>


<dt id="arg-root-edge">root_edge<a class="anchor" aria-label="anchor" href="#arg-root-edge"></a></dt>
<dd><p>logical Whether to plot root edge</p></dd>


<dt id="arg-raster">raster<a class="anchor" aria-label="anchor" href="#arg-raster"></a></dt>
<dd><p>logical Whether to raster images</p></dd>


<dt id="arg-show-phylo">show_phylo<a class="anchor" aria-label="anchor" href="#arg-show-phylo"></a></dt>
<dd><p>logical Whether to display phylogeny on y axis</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>ggplot panel</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">p</span> <span class="op">=</span> <span class="fu">plot_phylo_heatmap</span><span class="op">(</span></span></span>
<span class="r-in"><span>    gtree <span class="op">=</span> <span class="va">phylogeny_example</span>,</span></span>
<span class="r-in"><span>    joint_post <span class="op">=</span> <span class="va">joint_post_example</span>,</span></span>
<span class="r-in"><span>    segs_consensus <span class="op">=</span> <span class="va">segs_example</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

