<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Mouse tutorial • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Mouse tutorial">
<meta name="description" content="Haplotype-aware CNV analysis for mouse scRNA-seq">
<meta property="og:description" content="Haplotype-aware CNV analysis for mouse scRNA-seq">
<meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Preparing Inputs for Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Mouse tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/main/vignettes/mouse.Rmd" class="external-link"><code>vignettes/mouse.Rmd</code></a></small>
      <div class="d-none name"><code>mouse.Rmd</code></div>
    </div>

    
    
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@tGaoTeng">
<meta name="twitter:title" content="Numbat mouse tutorial">
<meta name="twitter:description" content="Haplotype-aware CNV analysis for mouse scRNA-seq">
<meta name="twitter:image" content="https://www.jax.org/-/media/jaxweb/images/news-and-insights/black-6-mama.jpg">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The mouse is a major model system for studying cancer and many other
areas of biology. According to the 10x <a href="https://www.10xgenomics.com/resources/publications?refinementList%5BproductGroups%5D%5B0%5D=Single%20Cell%20Gene%20Expression&amp;refinementList%5Btags%5D%5B0%5D=Cancer%20Research&amp;page=1" class="external-link">website</a>,
there are almost as many cancer single-cell studies on mouse subjects as
on humans.</p>
<p>Previously, we have shown that with prior haplotype information,
Numbat can detect allele-specific CNVs at a much higher accuracy <a href="#references">[1]</a>. Although we don’t usually know the haplotype
of a given human subject <em>a priori</em>, it can be “guessed”
statistically using <a href="https://en.wikipedia.org/wiki/Haplotype_estimation" class="external-link">population-based
phasing</a>. The resulting haplotypes are reasonably accurate within
short blocks (&lt;1Mb), but are subject to switch errors over a longer
range. The mouse is a special case because the haplotypes of most common
laboratory strains are fully known (thanks to the mouse genome projects
<a href="#references">[2,3]</a>). This means that all heterozygous SNPs
in a F1 hybrid mouse of two inbred strains are <em>fully phased</em> at
whole-genome level. In addition, the SNP density in a F1 hybrid mouse
can be many times higher than in humans <a href="#references">[2]</a>.
As we will see, these factors provide a huge advantage for
haplotype-aware CNV analysis.</p>
<p>The most updated version of Numbat (v1.2.0) now accepts (F1 hybrid)
mouse data. The input format is the same as usual (expression counts in
a matrix, allele counts in a long dataframe), but some bioinformatics is
required to prepare the phased allele counts. This tutorial will walk
through data prepration to analyze a mouse PDAC scRNA-seq dataset
provided by <a href="https://doi.org/10.1158/0008-5472.CAN-22-1742" class="external-link">Pitter et al</a> <a href="#references">[4]</a>.</p>
</div>
<div class="section level2">
<h2 id="preparing-input">Preparing input<a class="anchor" aria-label="anchor" href="#preparing-input"></a>
</h2>
<div class="section level3">
<h3 id="prepare-the-vcf">1. Prepare the VCF<a class="anchor" aria-label="anchor" href="#prepare-the-vcf"></a>
</h3>
<p>First, find out the parental strains of your F1 mouse. In the case of
Pitter et al, the mice are of C57BL/6 x Sv129 mixed genetic background.
This is convenient because the genome of C57BL/6 (aka <a href="https://www.jax.org/news-and-insights/2019/june/the-genome-of-eve-the-mouse" class="external-link">Eve</a>
the mouse) <em>is</em> the mm10 reference genome. This means that all
heterozygous SNPs in the F1 are the homozygous SNPs in Sv129. To create
the F1 VCF, we just need to take the Sv129 VCF and change the genotype
of homozygous SNPs (1/1) to heterozygous (1/0).</p>
<p>The VCFs of common lab mouse strains are available on <a href="http://hgdownload.soe.ucsc.edu/gbdb/mm10/mouseStrains/" class="external-link">UCSC
portal</a>. I downloaded the all-strain VCF
(<code>mgpV5MergedSNPsAlldbSNP142.vcf.gz</code>) and extracted the
passed homozygous sites for Sv129 using <code>bcftools</code>:</p>
<pre class="shell"><code>bcftools view mgpV5MergedSNPsAlldbSNP142.vcf.gz -c1 -s 129S1_SvImJ \
    -i "%FILTER='PASS'" | \
    bcftools annotate -Oz -x INFO,^FORMAT/GT &gt; 129S1_SvImJ.sites.vcf.gz</code></pre>
<p>We can then use <code>129S1_SvImJ.sites.vcf.gz</code> as the VCF for
our F1 mouse (except that all GT has to be changed to
<code>1/0</code>).</p>
<p><em>Note</em>: This is slightly more complicated if neither parental
strain is C57BL/6. One would have to find the SNPs that are homozygous
only in one parent but not both.</p>
</div>
<div class="section level3">
<h3 id="run-pileup">2. Run pileup<a class="anchor" aria-label="anchor" href="#run-pileup"></a>
</h3>
<p>For samples in this study, I downloaded raw FASTQs from the <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA861311" class="external-link">SRA</a> and
processed them by <code>cellranger</code>. Now we can generate
single-cell allele counts for the SNP positions in the above VCF using
<code>cellsnp-lite</code>.</p>
<pre class="shell"><code>cellsnp-lite \
    -s $cellrangerout/outs/possorted_genome_bam.bam \
    -b $cellrangerout/outs/filtered_feature_bc_matrix/barcodes.tsv.gz \
    -O ./pileup \
    -R ./129S1_SvImJ.sites.vcf.gz \
    -p 25 \
    --minMAF 0 \
    --minCOUNT 1</code></pre>
<p>where <code>$cellrangerout</code> is the cellranger output folder for
a specific sample.</p>
</div>
<div class="section level3">
<h3 id="prepare-allele-dataframe">3. Prepare allele dataframe<a class="anchor" aria-label="anchor" href="#prepare-allele-dataframe"></a>
</h3>
<p>Let’s now read in the pileup output in R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pu_dir</span> <span class="op">=</span> <span class="st">'./pileup'</span></span>
<span></span>
<span><span class="va">vcf_pu</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'{pu_dir}/cellSNP.base.vcf'</span><span class="op">)</span>, skip <span class="op">=</span> <span class="st">'#CHROM'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="va">`#CHROM`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>snp_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">CHROM</span>, <span class="va">POS</span>, <span class="va">REF</span>, <span class="va">ALT</span>, sep <span class="op">=</span> <span class="st">'_'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">CHROM</span>, <span class="st">'chr'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">CHROM</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">CHROM</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CHROM</span> <span class="op">!=</span> <span class="st">'X'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vcf_phased</span> <span class="op">=</span> <span class="va">vcf_pu</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>GT <span class="op">=</span> <span class="st">'1|0'</span>, cM <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>where the last line is because one of the parental genotypes is the
mm10 reference genome, all the alternative alleles in the VCF are
already <em>in phase</em>, so all GT should be <code>1|0</code>. In
addition, since there is no recombination, we set the genetic distance
<code>cM</code> to be a constant value for all sites.</p>
<p><em>Note</em>: Again, if neither parental strain is C57BL/6, one has
to assign GT to be <code>1|0</code> or <code>0|1</code> based on which
parent the variant came from.</p>
<p>Next, we can read in the allele counts and convert it into a long
dataframe.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pileup count matrices</span></span>
<span><span class="va">AD</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'{pu_dir}/cellSNP.tag.AD.mtx'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">DP</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'{pu_dir}/cellSNP.tag.DP.mtx'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">barcodes</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'{pu_dir}/cellSNP.samples.tsv'</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">V1</span></span>
<span></span>
<span><span class="co"># convert to dataframe</span></span>
<span><span class="va">DP</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">summary</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        cell <span class="op">=</span> <span class="va">barcodes</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,</span>
<span>        snp_id <span class="op">=</span> <span class="va">vcf_pu</span><span class="op">$</span><span class="va">snp_id</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">i</span>, <span class="op">-</span><span class="va">j</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>DP <span class="op">=</span> <span class="va">x</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">snp_id</span>, <span class="va">DP</span><span class="op">)</span></span>
<span></span>
<span><span class="va">AD</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">summary</span><span class="op">(</span><span class="va">AD</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        cell <span class="op">=</span> <span class="va">barcodes</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,</span>
<span>        snp_id <span class="op">=</span> <span class="va">vcf_pu</span><span class="op">$</span><span class="va">snp_id</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">i</span>, <span class="op">-</span><span class="va">j</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>AD <span class="op">=</span> <span class="va">x</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">snp_id</span>, <span class="va">AD</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df_allele</span> <span class="op">=</span> <span class="va">DP</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">AD</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell"</span>, <span class="st">"snp_id"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>AD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">AD</span><span class="op">)</span>, <span class="fl">0</span>, <span class="va">AD</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># attach genotype info</span></span>
<span><span class="va">df_allele</span> <span class="op">=</span> <span class="va">df_allele</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span></span>
<span>    <span class="va">vcf_phased</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">snp_id</span>, <span class="va">CHROM</span>, <span class="va">POS</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">GT</span>, <span class="va">cM</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">'snp_id'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-the-expression-reference">4. Prepare the expression reference<a class="anchor" aria-label="anchor" href="#prepare-the-expression-reference"></a>
</h3>
<p>Because these are pancreatic tumors (PDAC), I downloaded the
expression counts for a mouse normal pancreas scRNA-seq dataset (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE159343" class="external-link">GSE159343</a>),
ran clustering using <code>pagoda2</code>, and used it to construct the
reference expression profile (<code>ref_pancreas</code>).</p>
<pre><code><span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html" class="external-link">readMM</a></span><span class="op">(</span><span class="st">'~/GSE159343_RAW/GSM4826923_C57BL6J_matrix.mtx.gz'</span><span class="op">)</span></span>
<span><span class="va">cells</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">'~/GSE159343_RAW/GSM4826923_C57BL6J_cells.tsv.gz'</span>, header <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">V1</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">'~/GSE159343_RAW/GSM4826923_C57BL6J_genes.tsv.gz'</span>, header <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">V2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">genes</span></span>
<span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span></span>
<span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">rowsum</a></span><span class="op">(</span><span class="va">count_mat</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">count_mat</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span><span class="va">count_mat_ref</span> <span class="op">=</span> <span class="va">count_mat</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu">pagoda2</span><span class="fu">::</span><span class="fu">basicP2proc</span><span class="op">(</span><span class="va">count_mat_ref</span>, n.cores <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op">=</span> <span class="va">p2</span><span class="op">$</span><span class="va">clusters</span><span class="op">$</span><span class="va">PCA</span><span class="op">$</span><span class="va">multilevel</span></span>
<span></span>
<span><span class="va">ref_annot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>,</span>
<span>    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref_pancreas</span> <span class="op">=</span> <span class="fu">numbat</span><span class="fu">::</span><span class="fu"><a href="../reference/aggregate_counts.html">aggregate_counts</a></span><span class="op">(</span></span>
<span>    <span class="va">count_mat_ref</span>,</span>
<span>    <span class="va">ref_annot</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
<p>The expression count matrices for the tumor samples are prepared from
the cellranger output as usual.</p>
</div>
</div>
<div class="section level2">
<h2 id="run-numbat">Run Numbat<a class="anchor" aria-label="anchor" href="#run-numbat"></a>
</h2>
<p>Finally, let’s run CNV analysis using the prepared data. To let
Numbat know this is mouse data, we set the genome build as
<code>mm10</code> and set phase switch rate <code>nu = 0</code> to
disable phase switch (because the phasing is perfect!).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_numbat.html">run_numbat</a></span><span class="op">(</span></span>
<span>    <span class="va">count_mat_tumor</span>, </span>
<span>    <span class="va">ref_pancreas</span>, </span>
<span>    <span class="va">df_allele</span>,</span>
<span>    t <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span>    ncores <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    skip_nj <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    min_LLR <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    out_dir <span class="op">=</span> <span class="st">'./results'</span>,</span>
<span>    <span class="co"># mouse specific settings</span></span>
<span>    genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>    nu <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><em>Note</em>: For an admix with C57BL/6 that is not 50/50, you’d
create the VCF in the same way, and do a pileup also the same way. Since
this mouse won’t have all the alternative strain SNPs, you’d filter the
df_allele based on the bulk AF to only keep the het SNPs (VAF of ~50%)
after pileup. Then you can run Numbat with the default switch rate (nu =
1), since you’d expect recombination to create phase switches.</p>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>Let’s look at the results for sample KPT_062521 (SRR20462475), where
we see tumor subclones with distinct copy number profiles (the majority
of the events are copy-neutral LOH).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span> <span class="op">=</span> <span class="va"><a href="../reference/Numbat.html">Numbat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'./results/{sample}'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">8</span>, repr.plot.height <span class="op">=</span> <span class="fl">4</span>, repr.plot.res <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clone_pal</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="st">'gray'</span>, `2` <span class="op">=</span> <span class="st">"#E41A1C"</span>, `3` <span class="op">=</span> <span class="st">"#377EB8"</span>,</span>
<span>    `4` <span class="op">=</span> <span class="st">"#4DAF4A"</span>, `5` <span class="op">=</span> <span class="st">"#984EA3"</span>, `6` <span class="op">=</span> <span class="st">'bisque3'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nb</span><span class="op">$</span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span>pal_clone <span class="op">=</span> <span class="va">clone_pal</span>, clone_stack <span class="op">=</span> <span class="cn">TRUE</span>, p_min <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://i.imgur.com/UWuz9MT.png"></p>
<p>Let’s compare the copy number profiles of the two major subclones
(clone 4 and 5) in a pseudobulk HMM view.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">10</span>, repr.plot.height <span class="op">=</span> <span class="fl">4.5</span>, repr.plot.res <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nb</span><span class="op">$</span><span class="va">bulk_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CHROM</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span><span class="fu"><a href="../reference/plot_bulks.html">plot_bulks</a></span><span class="op">(</span>min_depth <span class="op">=</span> <span class="fl">15</span>, use_pos <span class="op">=</span> <span class="cn">F</span>, exp_limit <span class="op">=</span> <span class="fl">2.5</span>, min_LLR <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/tdtW7DY.png"></p>
<p>We see haplotype-specific differences with remarkable resolution. For
example, the subclonal copy number profiles for chr3 largely agree,
whereas chr4 and chr9 have distinct breakpoints and mirorred CNLOH
(chr4, middle segment). All alleles collapse to one side in a CNV
region, thanks to the perfect phasing.</p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Haplotype-aware CNV analysis is espcially powerful for model systems
(the mouse in particular) where the haplotypes are fully known. Luckily,
the logic in Numbat are written generically enough such that minimal
tweaking to the algorithm is required (simply setting phase switch rate
to be zero is sufficient, making it a special case). The main limitation
is that Numbat is only applicable to F1 hybrid mice with mixed genetic
backgrounds, whereas many mouse tumor models are conducted on a inbred
(pure) genetic background, which lacks heterozygous SNPs. Nonetheless,
this extension opens up many oppurtunities for integrative single-cell
analysis in a genetically manipulatable mammalian system.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Gao, T., Soldatov, R., Sarkar, H. et al. Haplotype-aware analysis
of somatic copy number variations from single-cell transcriptomes. Nat
Biotechnol (2022). <a href="https://doi.org/10.1038/s41587-022-01468-y" class="external-link uri">https://doi.org/10.1038/s41587-022-01468-y</a></p></li>
<li><p>Keane, T., Goodstadt, L., Danecek, P. et al. Mouse genomic
variation and its effect on phenotypes and gene regulation. Nature 477,
289–294 (2011). <a href="https://doi.org/10.1038/nature10413" class="external-link uri">https://doi.org/10.1038/nature10413</a></p></li>
<li><p>Lilue, J., Doran, A.G., Fiddes, I.T. et al. Sixteen diverse
laboratory mouse reference genomes define strain-specific haplotypes and
novel functional loci. Nat Genet 50, 1574–1583 (2018). <a href="https://doi.org/10.1038/s41588-018-0223-8" class="external-link uri">https://doi.org/10.1038/s41588-018-0223-8</a></p></li>
<li><p>Kenneth L. Pitter, et al. Systematic Comparison of Pancreatic
Ductal Adenocarcinoma Models Identifies a Conserved Highly Plastic Basal
Cell State. Cancer Res 1 October 2022; 82 (19): 3549–3560. <a href="https://doi.org/10.1158/0008-5472.CAN-22-1742" class="external-link uri">https://doi.org/10.1158/0008-5472.CAN-22-1742</a></p></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
