<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title> • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/main/vignettes/numbat.Rmd" class="external-link"><code>vignettes/numbat.Rmd</code></a></small>
      <div class="d-none name"><code>numbat.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="table-of-contents">Table of contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li>
<a href="#installation">Installation</a>
<ul>
<li><a href="#docker">Docker</a></li>
</ul>
</li>
<li><a href="#preparing-data">Preparing data</a></li>
<li>
<a href="#running-numbat">Running Numbat</a>
<ul>
<li><a href="#run-parameters">Run parameters</a></li>
<li><a href="#detecting-clonal-loh">Detecting clonal LOH</a></li>
<li><a href="#using-existing-cnv-calls">Using existing CNV
calls</a></li>
</ul>
</li>
<li><a href="#understanding-results">Understanding results</a></li>
<li>
<a href="#phasing-options">Phasing options</a>
<ul>
<li><a href="#using-prephased-snp-profiles">Using prephased SNP
profiles</a></li>
</ul>
</li>
<li><a href="#spatial-data">Spatial data</a></li>
<li><a href="#mouse-data">Mouse data</a></li>
<li><a href="#faq">FAQ</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>We now provide a ready-to-run Docker container that includes the
package and all prerequisites (see section <a href="#docker">Docker</a>). Alternatively, you can follow the
installation procedure below.</p>
<p>Numbat uses cellsnp-lite for generating SNP pileup data and eagle2
for phasing. Please follow their installation instructions and make sure
their binary executables can be found in your $PATH.</p>
<ol style="list-style-type: decimal">
<li><a href="https://github.com/single-cell-genetics/cellsnp-lite" class="external-link">cellsnp-lite</a></li>
<li><a href="https://alkesgroup.broadinstitute.org/Eagle/" class="external-link">eagle2</a></li>
<li><a href="http://www.htslib.org/" class="external-link">samtools</a></li>
</ol>
<p>Additionally, Numbat needs a common SNP VCF and phasing reference
panel. You can use the 1000 Genome reference below:</p>
<ol start="4" style="list-style-type: decimal">
<li>1000G SNP VCF</li>
</ol>
<pre><code># hg38
wget https://sourceforge.net/projects/cellsnp/files/SNPlist/genome1K.phase3.SNP_AF5e2.chr1toX.hg38.vcf.gz
# hg19
wget https://sourceforge.net/projects/cellsnp/files/SNPlist/genome1K.phase3.SNP_AF5e2.chr1toX.hg19.vcf.gz</code></pre>
<ol start="5" style="list-style-type: decimal">
<li>1000G Reference Panel (paste link in browser to download if
<code>wget</code> isn’t working)</li>
</ol>
<pre><code># hg38
wget http://pklab.med.harvard.edu/teng/data/1000G_hg38.zip
# hg19
wget http://pklab.med.harvard.edu/teng/data/1000G_hg19.zip</code></pre>
<p>You can install the numbat R package via CRAN:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'numbat'</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p>Alternatively, you can install the GitHub version:</p>
<pre><code><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"https://github.com/kharchenkolab/numbat"</span><span class="op">)</span></span></code></pre>
<div class="section level3">
<h3 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a>
</h3>
<p>We provide a ready-to-run Docker container that includes the Numbat R
package and all of its dependencies. You can launch it as follows:</p>
<pre><code>docker run -v /work:/mnt/mydata -it pkharchenkolab/numbat-rbase:latest /bin/bash</code></pre>
<p>where <code>/work</code> is a local data folder you would like to
access and write to. Within the container, <code>cellsnp-lite</code>,
<code>eagle2</code> and <code>samtools</code> are available in path and
the 1000 Genome SNP VCF and phasing panel files (hg38) are stored under
<code>/data</code>. You can also launch R and run interactive analysis
using Numbat (or run an R script).</p>
</div>
</div>
<div class="section level2">
<h2 id="preparing-data">Preparing data<a class="anchor" aria-label="anchor" href="#preparing-data"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Prepare the allele data. Run the preprocessing script
(<code>numbat/inst/bin/pileup_and_phase.R</code>) to count alleles and
phase SNPs.</li>
</ol>
<pre><code>usage: pileup_and_phase.R [-h] --label LABEL --samples SAMPLES --bams BAMS
                          [--barcodes BARCODES] --gmap GMAP [--eagle EAGLE]
                          --snpvcf SNPVCF --paneldir PANELDIR --outdir OUTDIR
                          --ncores NCORES [--UMItag UMITAG]
                          [--cellTAG CELLTAG] [--smartseq] [--bulk]

Run SNP pileup and phasing with 1000G

Arguments:
  -h, --help           show this help message and exit
  --label LABEL        Individual label. One per run.
  --samples SAMPLES    Sample name(s); comma delimited if multiple. 
                       All samples must belong to the same individual.
  --bams BAMS          BAM file(s); one per sample, comma delimited if multiple.
  --barcodes BARCODES  Cell barcode file(s); one per sample, 
                       comma delimited if multiple.
  --gmap GMAP          Path to genetic map provided by Eagle2 (e.g.
                       Eagle_v2.4.1/tables/genetic_map_hg38_withX.txt.gz)
  --eagle EAGLE        Path to Eagle2 binary file
  --snpvcf SNPVCF      SNP VCF for pileup
  --paneldir PANELDIR  Directory to phasing reference panel (BCF files)
  --outdir OUTDIR      Output directory
  --ncores NCORES      Number of cores
  --smartseq           Running with SMART-seq mode; Supply a txt file containing 
                       directories of BAM files to --bams and a txt file 
                       containing cell names to --barcodes (each entry on its 
                       own line for both; ordering must match).</code></pre>
<p>For example, within the Numbat Docker <a href="#docker">container</a>
you can run the preprocessing script like this:</p>
<pre><code>Rscript /numbat/inst/bin/pileup_and_phase.R \
    --label {sample} \
    --samples {sample} \
    --bams /mnt/mydata/{sample}.bam \
    --barcodes /mnt/mydata/{sample}_barcodes.tsv \
    --outdir /mnt/mydata/{sample} \
    --gmap /Eagle_v2.4.1/tables/genetic_map_hg38_withX.txt.gz \
    --snpvcf /data/genome1K.phase3.SNP_AF5e2.chr1toX.hg38.vcf \
    --paneldir /data/1000G_hg38 \
    --ncores ncores</code></pre>
<p><strong>Important</strong>: If your 10x bam is <em>multiplexed</em>
(containing cells from multiple individuals), please only provide cell
barcodes for a single individual in each genotyping run.</p>
<p>This will produce a file named
<code>{sample}_allele_counts.tsv.gz</code> under the specified output
directory, which contains cell-level phased allele counts. It contains
columns <code>cell</code>, <code>snp_id</code>, <code>CHROM</code>,
<code>POS</code>, <code>cM</code> (genetic distance in centimorgan),
<code>REF</code>, <code>ALT</code>, <code>AD</code> (ALT allele count),
<code>DP</code> (total allele count), <code>GT</code> (phased genotype),
<code>gene</code>. If multiple samples from the same individual were
provided, there will be one allele count file for each sample. Other
outputs include phased vcfs under <code>phasing/</code> folder and raw
pileup counts under <code>pileup/</code>.</p>
<ol start="2" style="list-style-type: decimal">
<li><p>Prepare the expression data. Numbat takes a gene by cell integer
UMI count matrix as input. You can directly use results from upstream
transcriptome quantification pipelines such as 10x CellRanger.</p></li>
<li><p>Prepare the expression reference, which is a gene by cell type
matrix of normalized expression values (raw gene counts divided by total
counts). For a quick start, you may use a our HCA collection
(<code>ref_hca</code>) that ships with the package. If you have matched
normal cells (ideally, of various cell type) from the same patient or
dataset and would like to make your own references, you may use this
utility function:</p></li>
</ol>
<pre><code><span><span class="co"># count_mat is a gene x cell raw count matrices</span></span>
<span><span class="co"># cell_annot is a dataframe with columns "cell" and "group"</span></span>
<span><span class="va">ref_internal</span> <span class="op">=</span> <span class="fu"><a href="../reference/aggregate_counts.html">aggregate_counts</a></span><span class="op">(</span><span class="va">count_mat</span>, <span class="va">cell_annot</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="running-numbat">Running Numbat<a class="anchor" aria-label="anchor" href="#running-numbat"></a>
</h2>
<p>In this example (ATC2 from <a href="https://www.nature.com/articles/s41587-020-00795-2" class="external-link">Gao et
al</a>), the gene expression count matrix and allele dataframe are
already prepared for you.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kharchenkolab/numbat/" class="external-link">numbat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_mat_ATC2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">'http://pklab.med.harvard.edu/teng/data/count_mat_ATC2.rds'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_allele_ATC2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">'http://pklab.med.harvard.edu/teng/data/df_allele_ATC2.rds'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run</span></span>
<span><span class="va">out</span> <span class="op">=</span> <span class="fu"><a href="../reference/run_numbat.html">run_numbat</a></span><span class="op">(</span></span>
<span>    <span class="va">count_mat_ATC2</span>, <span class="co"># gene x cell integer UMI count matrix </span></span>
<span>    <span class="va">ref_hca</span>, <span class="co"># reference expression profile, a gene x cell type normalized expression level matrix</span></span>
<span>    <span class="va">df_allele_ATC2</span>, <span class="co"># allele dataframe generated by pileup_and_phase script</span></span>
<span>    genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>    t <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span>    ncores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    out_dir <span class="op">=</span> <span class="st">'./test'</span></span>
<span><span class="op">)</span></span></code></pre>
<p><strong>Note</strong>: If you wish to use your own custom reference,
please use the <code>aggregate_counts</code> function as per the example
in <a href="#preparing-data">preparing data</a>. You do not need to
include the reference cells in <code>count_mat</code> or
<code>df_allele</code>; only provide them as
<code>lambdas_ref</code>.</p>
<div class="section level3">
<h3 id="run-parameters">Run parameters<a class="anchor" aria-label="anchor" href="#run-parameters"></a>
</h3>
<p>There are a few parameters you can consider tuning to a specific
dataset.</p>
<p><em>CNV detection</em></p>
<ul>
<li>
<code>t</code>: the transition probability used in the HMM. A lower
<code>t</code> is more appropriate for tumors with more complex copy
number landscapes (from which you can expect more breakpoints) and is
sometimes better for detecting subclonal events. A higher <code>t</code>
is more effective for controlling false-positive rates of CNV
calls.</li>
<li>
<code>gamma</code>: Overdispersion in allele counts (default: 20).
For 10x data, 20 is recommended. Non-UMI protocols (e.g. SMART-Seq)
usually produce noisier allele data, and a smaller value of gamma is
recommended (e.g. 5).</li>
<li>
<code>min_cells</code>: minimum number of cells for which an
pseudobulk HMM will be run. If the allele coverage per cell is very
sparse for a dataset, then I would consider setting this threshold to be
higher.</li>
<li>
<code>multi_allelic</code>: Whether to enable calling of
multiallelic CNVs</li>
</ul>
<p><em>CNV filtering</em></p>
<ul>
<li>
<code>min_LLR</code>: minimum log-likelihood ratio threshold to
filter CNVs (default: 5). To ensure quality of phylogeny inference, we
only use confident CNVs to reconstruct the phylogeny.</li>
<li>
<code>max_entropy</code>: another criteria that we use to filter
CNVs before phylogeny construction (default: 0.5). The entropy of the
binary posterior quantifies the uncertainty of an event across single
cells. The value can be from 0 to 1 where 1 is the least stringent.</li>
</ul>
<p><em>Phylogeny</em></p>
<ul>
<li>
<code>n_cut</code>: Number of cuts on the phylogeny to define
subclones. Note that <code>n</code> cuts should result in
<code>n+1</code> clones (the top-level normal diploid clone is always
included). Alternatively, one can specify <code>max_cost</code> or
<code>tau</code> to control the number of subclones.</li>
<li>
<code>max_cost</code>: Likelihood threshold to collapse internal
branches. Higher <code>max_cost</code> generally leads to fewer
clones.</li>
<li>
<code>tau</code>: Stringency to simplify the mutational history.
Basically sets <code>max_cost</code> as <code>tau</code> times the
number of cells.</li>
</ul>
<p><em>Iterative optimization</em></p>
<ul>
<li>
<code>init_k</code>: initial number of subclusters to use for the
<code>hclust</code> initialization. Numbat by default uses hierarchical
clustering (<code>hclust</code>) of smoothed expression values to
approximate an initial phylogeny. This will cut the initial tree into k
clusters. More clusters means more resolution at the initial stage for
subclonal CNV detection. By default, we set init_k to be 3.</li>
<li>
<code>max_iter</code>: maximum number of iterations. Numbat
iteratively optimizes the phylogeny and copy number estimations. In
practice, we find that results after 2 iterations are usually
stable.<br>
</li>
<li>
<code>check_convergence</code>: stop iterating if the results have
converged (based on consensus CNV segments).</li>
</ul>
<p><em>Parallelization</em></p>
<ul>
<li>
<code>ncores</code>: number of cores to use for single-cell CNV
testing</li>
<li>
<code>ncores_nni</code>: number of cores to use for phylogeny
inference</li>
</ul>
<p><em>Other run modes</em></p>
<ul>
<li>
<code>call_clonal_loh</code>: whether to call clonal LOH regions.
Recommended for samples with high purity without matched normal cells.
See <a href="#detecting-clonal-loh">Detecting clonal LOH</a> for
details.</li>
<li>
<code>segs_consensus_fix</code>: a dataframe of pre-determined
consensus CNVs (for example, from bulk WGS/WES analysis) that will stay
fixed during analysis. See <a href="#using-existing-cnv-calls">Using
existing CNV calls</a> for details.</li>
</ul>
</div>
<div class="section level3">
<h3 id="detecting-clonal-loh">Detecting clonal LOH<a class="anchor" aria-label="anchor" href="#detecting-clonal-loh"></a>
</h3>
<p>In samples with high tumor purity (e.g., tumor cell lines) without
matched normal cells, heterozygous SNPs are challenging to identify in
regions of LoH, leading to decreased power of CNV detection. Regions of
clonal LoH have decreased SNP density and can be identified by a
specialized HMM (see <code>detect_clonal_loh</code> in function
reference). You can set <code>call_clonal_loh = TRUE</code> to
automatically identify and exclude regions of clonal deletions/LoH in
the main workflow. Alternatively, you can manually run
<code>detect_clonal_loh</code> and provide the detected segments as a
dataframe (with columns <code>CHROM</code>, <code>seg</code>,
<code>seg_start</code>, <code>seg_end</code>) via the
<code>segs_loh</code> argument.</p>
<p>In addition, if you have matched DNA data (e.g. WGS/WES) of the same
individual, you can directly use the SNP profile genotyped from DNA,
which should avoid the above problem. See <a href="#using-prephased-snp-profiles">Using prephased SNP profiles</a>
for details.</p>
</div>
<div class="section level3">
<h3 id="using-existing-cnv-calls">Using existing CNV calls<a class="anchor" aria-label="anchor" href="#using-existing-cnv-calls"></a>
</h3>
<p>Sometimes users already have CNV calls from bulk WGS, WES, or array
analysis. In this case, you can supply the existing CNV profile via
<code>segs_consensus_fix</code> parameter to fix the CNV boundaries and
states. To do so, you may provide a dataframe with the following
columns:</p>
<ul>
<li>
<code>CHROM</code>: integer; chromosome (1-22)</li>
<li>
<code>seg</code>: character; segment ID (e.g. 1a, 1b, 2a, 2b,
etc.)</li>
<li>
<code>seg_start</code>: integer; segment start position</li>
<li>
<code>seg_end</code>: integer; segment end position</li>
<li>
<code>cnv_state</code>: character; copy number state (neu, del, amp,
loh, bamp, bdel)</li>
</ul>
<p>Please note that diploid segments (<code>cnv_state = "neu"</code>)
should also be included (i.e. <code>segs_consensus_fix</code> should be
a complete copy number profile including all chromosomes).</p>
</div>
</div>
<div class="section level2">
<h2 id="understanding-results">Understanding results<a class="anchor" aria-label="anchor" href="#understanding-results"></a>
</h2>
<p>A detailed vignette on how to interpret and visualize Numbat results
is available:</p>
<ul>
<li><a href="https://kharchenkolab.github.io/numbat/articles/results.html">Interpreting
Numbat results</a></li>
</ul>
<p>For a full list of output files and column descriptions, please
see:</p>
<ul>
<li><a href="https://kharchenkolab.github.io/numbat/articles/descriptions.html">Output
descriptions</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="phasing-options">Phasing options<a class="anchor" aria-label="anchor" href="#phasing-options"></a>
</h2>
<p>The CNV detection power of Numbat can be further enhanced by
conducting population-based phasing with a larger and more diverse
reference panel (i.e. reference haplotypes). The default pipeline above
uses the <a href="https://www.internationalgenome.org" class="external-link">1000 Genome</a>
panel, which contains genotypes from 2,504 individuals. Larger reference
panels include:</p>
<ul>
<li><p>gnomAD HGDP + 1KG panel (n=4,099). You can download the reference
files using <a href="https://cloud.google.com/storage/docs/gsutil" class="external-link">gsutil</a>:
<code>gs://gcp-public-data--gnomad/resources/hgdp_1kg/phased_haplotypes</code>.</p></li>
<li><p>TOPMed panel (n=97,256). You can upload your VCF to the <a href="https://imputation.biodatacatalyst.nhlbi.nih.gov" class="external-link">TOPMed
imputation server</a>.</p></li>
</ul>
<div class="section level3">
<h3 id="using-prephased-snp-profiles">Using prephased SNP profiles<a class="anchor" aria-label="anchor" href="#using-prephased-snp-profiles"></a>
</h3>
<p>Using DNA-derived genotype information is another way to improve SNP
density and phasing. If you have SNP calls from DNA genotyping
(e.g. WGS/WES), you can first perform phasing on the DNA-derived VCF.
Then run <code>cellsnp-lite</code> on scRNA-seq BAMs against the
DNA-derived VCF to generate allele counts (only include heterozygous
SNPs). Finally, merge the phased GT fields (from phased DNA-derived VCF)
with the obtained allele counts to produce an allele dataframe in the
format of <code>df_allele</code> (see section <a href="#preparing-data">Preparing data</a>).</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-data">Spatial data<a class="anchor" aria-label="anchor" href="#spatial-data"></a>
</h2>
<p>Numbat can also be used to analyze spatial transcriptomics (e.g. 10x
Visium) data. Please see the <a href="https://kharchenkolab.github.io/numbat/articles/spatial-rna.html">spatial
tutorial</a> for details.</p>
</div>
<div class="section level2">
<h2 id="mouse-data">Mouse data<a class="anchor" aria-label="anchor" href="#mouse-data"></a>
</h2>
<p>Numbat (v1.2.0) now works for a F1 hybrid mouse bred from two known
lab mice strains. More details are available in the <a href="https://kharchenkolab.github.io/numbat/articles/mouse.html">mouse
tutorial</a>.</p>
</div>
<div class="section level2">
<h2 id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h2>
<p><strong>Q</strong>: What is the expression reference for and how does
it impact analysis?</p>
<p><strong>A</strong>: The expression reference
(<code>lambdas_ref</code>) is used to regress out any cell-type-specific
expression differences that are unrelated to CNVs. It is used in both
single-cell testing and pseudobulk CNV calling using HMM. Ideally,
expression reference should be created from a normal (non-tumor) dataset
closely matching the biological and technical conditions of the target
dataset (e.g., same tissue types, cell types, sequencing platforms). A
closely matching reference can help reduce the noise in the
expression-based CNV signal (logFC) during pseudobulk and single-cell
analysis.</p>
<p><strong>Q</strong>: We expect a certain CNV to be in a particular
sample, but it is not detected by Numbat.</p>
<p><strong>A</strong>: In general, there are three scenarios that a CNV
is not called.</p>
<ol style="list-style-type: decimal">
<li><p>The CNV is not called in pseudobulk HMM. You can check
<code>bulk_subtrees_*.png</code> to see if the event was called. If it
was not, I would suggest trying to find a better expression reference to
reduce the noise in logFC, or changing the parameters used to configure
the HMM (<code>gamma</code>, <code>t</code>, etc). Sometimes the CNV is
very subclonal and was not isolated as a pseudobulk by the initial
clustering. You can see if this was the case by looking at the
<code>exp_roll_clust.png</code> and if so, try increasing
<code>k_init</code>.</p></li>
<li><p>CNV is called in pseudobulks but did not appear in
<code>segs_consensus_*.tsv</code> and <code>joint_post_*tsv</code>. This
is because the event is filtered out due to low LLR. Lowering
<code>min_LLR</code> threshold will help.</p></li>
<li><p>CNV is called in pseudobulks, is present in
<code>segs_consensus_*.tsv</code> and <code>joint_post_*tsv</code>, but
did not appear in phylogeny. This is because the event is filtered out
due to high entropy (weak evidence) in single cells. Raising
<code>max_entropy</code> threshold will help. You can check the entropy
of specific events in <code>joint_post_*tsv</code> in the column
<code>avg_entropy</code>.</p></li>
</ol>
<p><strong>Q</strong>: Numbat predicts most of the genome to be
amplified, and there are gaps in the pseudobulk allele profile with no
heterozygous SNPs.</p>
<p><strong>A</strong>: This is a sign that the sample consists mostly of
tumor (with very few normal cells), and clonal deletions are present.
The heterozygous SNPs cannot be identified in these regions, therefore
leaving gaps. You can identify these clonally deleted regions via the
SNP-density HMM prior to running the main <code>run_numbat</code>
workflow (see section <a href="#detecting-clonal-loh">Detecting Clonal
LOH</a>).</p>
<p><strong>Q</strong>: There seems to be a global baseline shift in the
CNV profile produced by Numbat as compared to my other CNV callers;
specifically, Numbat calls gains for the segments that appear to be
neutral in other analyses, and neutral for segments that appear to be
losses.</p>
<p><strong>A</strong>: Many existing methods (e.g. InferCNV/CopyKAT)
infer copy number variations relative to the median ploidy, which can
dilute signals of aberrant regions or mistake neutral regions for
aberrant due to baseline shifts caused by hyperdiploidy or hypodiploidy.
Instead, Numbat first tries to identify diploid regions based on allele
evidence (balanced allelic frequencies), and uses these regions as
baseline for CNV calling. For more information, please refer to the
publications describing <a href="https://www.nature.com/articles/s41587-022-01468-y" class="external-link">Numbat</a> and
<a href="https://academic.oup.com/nar/article/44/16/e131/2460163?login=false" class="external-link">FACETS</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
