<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Preparing Inputs for Numbat-multiome • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Preparing Inputs for Numbat-multiome">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/numbat-multiome.html">Preparing Inputs for Numbat-multiome</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Preparing Inputs for Numbat-multiome</h1>
                        <h4 data-toc-skip class="author">Ruitong Li,
Tina Keshavarzian and Teng Gao</h4>
            
            <h4 data-toc-skip class="date">2025-10-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/main/vignettes/numbat-multiome.Rmd" class="external-link"><code>vignettes/numbat-multiome.Rmd</code></a></small>
      <div class="d-none name"><code>numbat-multiome.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a guide to preparing input data for running
Numbat-multiome, which extends the capabilities of Numbat to perform CNV
inference from accessible chromatin profiles (through scATAC-seq data)
alone, as well as mixed modality (separate scRNA-seq and scATAC-seq
datasets of the same sample) in an integrated manner. Note that joint
multiome assays such as the <a href="https://www.10xgenomics.com/products/epi-multiome" class="external-link">10x
Multiome</a> (direct linkage of gene expression and open chromatin
readouts from the same nucleus) is not yet fully supported (currently
under development).</p>
<p>The core inputs required are:</p>
<ul>
<li>
<strong>Reference feature matrix</strong>: feature (gene or genomic
bin) by cell‑type normalized count matrix.</li>
<li>
<strong>GTF file</strong>: annotation mapping features to genomic
locations, used to connect SNPs and features.</li>
<li>
<strong>Feature count matrix</strong>: raw count matrix (RNA:
cell-by-gene/bin matrix filled with integer counts of UMI barcodes;
ATAC: cell-by-bin matrix filled with aggregated integer counts of
individual Tn5-accessible chromatin fragments, defined by unique
combination of chromosome:start-end genomic coordinates).</li>
<li>
<strong>Phased SNP genotype data</strong>: single‑cell phased
genotype of SNP (TSV.GZ).</li>
</ul>
<p>This document covers two preparation modes:</p>
<ul>
<li>
<code>Combined-bin</code>: Joint RNA and ATAC CNV analysis using
binned inputs.</li>
<li>
<code>ATAC-bin</code>: ATAC‑based CNV analysis using binned
accessibility data.</li>
</ul>
</div>
<div class="section level2">
<h2 id="combined-bin-combined-rna-and-atac-cnv-analysis">Combined-bin: Combined RNA and ATAC CNV analysis<a class="anchor" aria-label="anchor" href="#combined-bin-combined-rna-and-atac-cnv-analysis"></a>
</h2>
<div class="section level3">
<h3 id="prepare-snp-allele-data-for-both-modalities">1. Prepare SNP allele data for both modalities<a class="anchor" aria-label="anchor" href="#prepare-snp-allele-data-for-both-modalities"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="va">nc</span><span class="op">=</span>8 <span class="co"># number of cores to use</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="va">sample</span><span class="op">=</span><span class="st">"MM1"</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">### default to path in numbat image</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="va">phase_panel</span><span class="op">=</span><span class="st">"/data/1000G_hg38"</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="va">vcf_genome1k</span><span class="op">=</span><span class="st">"/data/genome1K.phase3.SNP_AF5e2.chr1toX.hg38.vcf"</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="va">gma_gz</span><span class="op">=</span><span class="st">"/Eagle_v2.4.1/tables/genetic_map_hg38_withX.txt.gz"</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/pileup_and_phase.R <span class="dt">\</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">--label</span> <span class="va">${sample}</span> <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="at">--samples</span> <span class="va">${sample}</span>_rna,<span class="va">${sample}</span>_atac <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">--bams</span> <span class="va">${sample}</span>_rna.bam,<span class="va">${sample}</span>_atac.bam <span class="dt">\</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="at">--barcodes</span> <span class="va">${sample}</span>_rna_barcodes.tsv,<span class="va">${sample}</span>_atac_barcodes.tsv <span class="dt">\</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>  <span class="at">--gmap</span> <span class="va">${gma_gz}</span> <span class="dt">\</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>  <span class="at">--snpvcf</span> <span class="va">${vcf_genome1k}</span> <span class="dt">\</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>  <span class="at">--paneldir</span> <span class="va">${phase_panel}</span> <span class="dt">\</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>  <span class="at">--ncores</span> <span class="va">${nc}</span> <span class="dt">\</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>  <span class="at">--cellTAG</span> CB <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>  <span class="at">--UMItag</span> Auto,None <span class="dt">\</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>  <span class="at">--outdir</span> <span class="va">${outdir}</span></span></code></pre></div>
<p>The resulting RNA and ATAC allele count dataframes are saved in the
<code>outdir</code> directory, as
<code>${sample}_rna_allele_counts.tsv.gz</code> and
<code>${sample}_atac_allele_counts.tsv.gz</code> respectively. You can
combine them into a single combined dataframe using
<code><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">dplyr::bind_rows()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="prepare-binned-rna-atac-inputs">2. Prepare binned RNA &amp; ATAC inputs<a class="anchor" aria-label="anchor" href="#prepare-binned-rna-atac-inputs"></a>
</h3>
<div class="section level4">
<h4 id="generate-gene-to-bin-mapping">Generate gene-to-bin mapping<a class="anchor" aria-label="anchor" href="#generate-gene-to-bin-mapping"></a>
</h4>
<p>You can download a premade genomic bin object here: <a href="https://raw.githubusercontent.com/kharchenkolab/numbat/main/inst/extdata/var220kb.rds" class="external-link">var220kb.rds</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="va">binGR</span><span class="op">=</span><span class="st">"var220kb.rds"</span> <span class="co"># path to grange object containing genomic bins</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="va">gtfF</span><span class="op">=</span><span class="st">"gtf_hg38.gtf"</span> <span class="co"># any version of gtf files you use or hg38, hg19 or mm10</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_gene_binned_intersections.R <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="at">--numbatGTFname</span> <span class="va">${gtfF}</span> <span class="dt">\ </span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="ex">--binGR</span> <span class="va">${binGR}</span> <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">--outfile</span> gene2bin_map.csv</span></code></pre></div>
<p>Here, you can either supply a gtf-format file if you have a
customized version, if you prefer using the default numbat gtf files,
you can simply use <code>hg38</code>, <code>hg19</code> or
<code>mm10</code> as the <code>numbatGTFname</code> argument.</p>
</div>
<div class="section level4">
<h4 id="generate-binned-expression-counts">Generate binned expression counts<a class="anchor" aria-label="anchor" href="#generate-binned-expression-counts"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_binned_rna.R <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="at">--rnaCountsFile</span> <span class="va">${sample}</span>_seu.rds <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">--outFile</span> <span class="va">${sample}</span>/<span class="va">${sample}</span>_rna_bin.rds <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">--barcodesKeep</span> <span class="va">${sample}</span>_barcodes.tsv <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">--geneBinMapCSVFile</span> gene2bin_map.csv</span></code></pre></div>
<table class="table">
<colgroup>
<col width="13%">
<col width="86%">
</colgroup>
<thead><tr class="header">
<th align="left">Argument</th>
<th align="left">Accepted types / format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>--rnaCountsFile</code></td>
<td align="left">
<strong>.rds</strong> file containing either:<br> • a
<strong>Seurat</strong> object (extracts
<code>@assays[[1]]@counts</code>)<br> • a raw counts matrix
(<code>dgCMatrix</code> etc.)<br> • <strong>.h5</strong> file from 10x
(uses <code>Read10X_h5</code>)</td>
</tr>
<tr class="even">
<td align="left"><code>--geneBinMapCSVFile</code></td>
<td align="left">CSV with <strong>at least</strong> columns
<code>gene_name</code> and <code>bingrp</code> (e.g. output of your
<code>get_gene_binned_intersections.R</code>).</td>
</tr>
<tr class="odd">
<td align="left"><code>--barcodesKeep</code></td>
<td align="left">- A <strong>TSV/CSV</strong> text file with one barcode
per line (first column).<br>- Or an R data.frame containing one column
of barcodes.</td>
</tr>
<tr class="even">
<td align="left"><code>--outFile</code></td>
<td align="left">Path for saving the binned-by-cell matrix (always as an
RDS of a sparse <code>dgCMatrix</code>).</td>
</tr>
<tr class="odd">
<td align="left"><code>--generateAggRef</code></td>
<td align="left">
<em>(flag)</em> if included, also creates a
Numbat-style aggregated reference matrix.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="generate-binned-atac-counts">Generate binned ATAC counts<a class="anchor" aria-label="anchor" href="#generate-binned-atac-counts"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="va">binGR</span><span class="op">=</span><span class="st">"var220kb.rds"</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_binned_atac.R <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">--CB</span> <span class="va">${sample}</span>_atac_barcodes.tsv <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">--frag</span> <span class="va">${sample}</span>_fragments.tsv.gz <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">--binGR</span> <span class="va">${binGR}</span> <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">--outFile</span> <span class="va">${sample}</span>/<span class="va">${sample}</span>_atac_bin.rds</span></code></pre></div>
<table class="table">
<colgroup>
<col width="17%">
<col width="82%">
</colgroup>
<thead><tr class="header">
<th align="left">Argument</th>
<th align="left">Accepted types / format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>--CB</code></td>
<td align="left">TSV/CSV text file listing <strong>cell
barcodes</strong> (one per line, first column).</td>
</tr>
<tr class="even">
<td align="left"><code>--frag</code></td>
<td align="left">Fragments file in <strong>BED-style</strong> TSV or
<code>.gz</code>, with columns: chrom, start, end, barcode,
duplicate-count.</td>
</tr>
<tr class="odd">
<td align="left"><code>--binGR</code></td>
<td align="left">RDS file containing a <strong>GRanges</strong> object
of genomic bins (e.g. <code>var220kb.rds</code>).</td>
</tr>
<tr class="even">
<td align="left"><code>--outFile</code></td>
<td align="left">Output path—if <code>.tsv</code>, writes a
tab-delimited table; if <code>.rds</code>, saves as an RDS matrix.</td>
</tr>
<tr class="odd">
<td align="left"><code>--generateAggRef</code></td>
<td align="left">
<em>(flag)</em> if included, also runs Numbat’s
<code>aggregate_counts</code> on the bin×cell matrix and writes that
result.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="generate-combined-count-matrix">Generate combined count matrix<a class="anchor" aria-label="anchor" href="#generate-combined-count-matrix"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{numbat_repo}/inst/bin/input_prep.R"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="fu">binCnt</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"${sample}/{sample}_rna_bin.rds"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"${sample}/{sample}_atac_bin.rds"</span><span class="op">)</span><span class="op">)</span>,seed<span class="op">=</span><span class="fl">123</span>,maxCB<span class="op">=</span><span class="fl">10000</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{sample}/{sample}_comb_bincnt.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can load in an interactive R session in your terminal or open
other IDEs to apply function <code>binCnt()</code> which combines the
binned RNA and ATAC count matrices into a single matrix which also did a
subsampling of cells to a maximum of 10,000 cells per modality, you can
specify the seed and maximum allowed cell number in the parameters.</p>
</div>
</div>
<div class="section level3">
<h3 id="generate-combined-reference">3. Generate combined reference<a class="anchor" aria-label="anchor" href="#generate-combined-reference"></a>
</h3>
<p>We generated binned counts from scRNA-seq and scATAC-seq data of
normal samples.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="va">binGR</span><span class="op">=</span><span class="st">"var220kb.rds"</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="va">refsample</span><span class="op">=</span><span class="st">"normal1"</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_binned_rna.R <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">--rnaCountsFile</span> ref_seu.rds <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">--outFile</span> Reference/lambdas_RNA_bincnt.rds <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">--barcodesKeep</span> ref_barcodes.tsv <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">--geneBinMapCSVFile</span> gene2bin_map.csv <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="at">--generateAggRef</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  </span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_binned_atac.R <span class="dt">\</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>  <span class="at">--CB</span> <span class="va">${refsample}</span>_atac_barcodes.tsv <span class="dt">\</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  <span class="at">--frag</span> <span class="va">${refsample}</span>_fragments.tsv.gz <span class="dt">\</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  <span class="at">--binGR</span> <span class="va">$binGR</span> <span class="dt">\</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  <span class="at">--outFile</span>  Reference/lambdas_RNA_bincnt.rds <span class="dt">\</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>  <span class="at">--generateAggRef</span></span></code></pre></div>
<p>If you have relatively high number of cells for multiple normal
samples, you may consider running <code>agg_refs</code>, a wrapper
function of the original <code>aggrerate_count()</code> function in
numbat but with subsampling max number of cells from each normal
sample.</p>
<p>First we can set up an array of normal samples to iterate through
bash array</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Define array of normal sample names</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="va">normal_samples</span><span class="op">=</span><span class="va">(</span><span class="st">"normal1"</span> <span class="st">"normal2"</span> <span class="st">"normal3"</span><span class="va">)</span>  <span class="co"># &lt;-- Replace with your sample names</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># Path to shared binGR file</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="va">binGR</span><span class="op">=</span><span class="st">"var220kb.rds"</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># Loop through each sample</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="cf">for</span> refsample <span class="kw">in</span> <span class="st">"</span><span class="va">${normal_samples</span><span class="op">[@]</span><span class="va">}</span><span class="st">"</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"Processing </span><span class="va">$refsample</span><span class="st">..."</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>  <span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/get_binned_atac.R <span class="dt">\</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>    <span class="at">--CB</span> <span class="st">"</span><span class="va">${refsample}</span><span class="st">_atac_barcodes.tsv"</span> <span class="dt">\</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>    <span class="at">--frag</span> <span class="st">"</span><span class="va">${refsample}</span><span class="st">_fragments.tsv.gz"</span> <span class="dt">\</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>    <span class="at">--binGR</span> <span class="st">"</span><span class="va">$binGR</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>    <span class="at">--outFile</span> <span class="st">"</span><span class="va">${refsample}</span><span class="st">/</span><span class="va">${refsample}</span><span class="st">_atac_bin.rds"</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>Specify <code>refsamples</code> as a vector of normal sample names,
e.g., <code>refsamples &lt;- c("normal1", "normal2", "normal3")</code>.
Then run the following R code to aggregate the references:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{numbat_repo}/inst/bin/input_prep.R"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ref_atac</span> <span class="op">&lt;-</span> <span class="fu">agg_refs</span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">refsamples</span>,<span class="st">"/"</span>,<span class="va">refsamples</span>,<span class="st">"_atac_bin.rds"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">set_names</span><span class="op">(</span><span class="va">refsamples</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="st">"Reference/lambdas_ATAC_bincnt.rds"</span><span class="op">)</span></span></code></pre></div>
<p>After separately generating binned RNA and ATAC reference, we can
combine them into a single reference.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Reference/lambdas_RNA_bincnt.rds"</span><span class="op">)</span></span>
<span><span class="va">ref_atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Reference/lambdas_ATAC_bincnt.rds"</span><span class="op">)</span></span>
<span><span class="va">shared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ref_rna</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ref_atac</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ref_comb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ref_rna</span><span class="op">[</span><span class="va">shared</span>, <span class="op">]</span>, <span class="va">ref_atac</span><span class="op">[</span><span class="va">shared</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ref_comb</span>, <span class="st">"Reference/lambdas_comb_bincnt.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-numbat-inference-with-prepared-inputs">Run numbat inference with prepared inputs<a class="anchor" aria-label="anchor" href="#run-numbat-inference-with-prepared-inputs"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="va">binGR</span><span class="op">=</span><span class="st">"var220kb.rds"</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="va">parL</span><span class="op">=</span><span class="st">"par_numbatm.rds"</span> <span class="co"># a list of any run_numbat parameters</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/run_numbat_multiome.R  <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>            <span class="at">--countmat</span> <span class="va">${sample}</span>/<span class="va">${sample}</span>_comb_bincnt.rds <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>            <span class="at">--alleledf</span> <span class="va">${sample}</span>/<span class="va">${sample}</span>_comb_allele_counts.tsv.gz <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>            <span class="at">--out_dir</span> <span class="va">${sample}</span>/paired/ <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>            <span class="at">--ref</span> Reference/lambdas_comb_bincnt.rds <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>            <span class="at">--gtf</span>  <span class="va">${binGR}</span><span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>            <span class="at">--parL</span> <span class="va">${parL}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="atac-bin-and-rna-bin-modes">ATAC-bin and RNA-bin modes<a class="anchor" aria-label="anchor" href="#atac-bin-and-rna-bin-modes"></a>
</h2>
<p>If you want to run Numbat-multiome in RNA bin and ATAC bin modes, you
can simply use <code>lambdas_RNA_bincnt.rds</code> and
<code>lambdas_ATAC_bincnt.rds</code> as the reference files.
<code>${sample}_rna_bin.rds</code> and
<code>${sample}_atac_bin.tsv</code> as the count matrix.</p>
<div class="section level3">
<h3 id="atac-bin-mode">ATAC-bin mode<a class="anchor" aria-label="anchor" href="#atac-bin-mode"></a>
</h3>
<p>In <strong>ATAC-bin</strong> mode you perform CNV inference using
only the binned ATAC counts and the ATAC reference.</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Inputs</strong></p>
<ul>
<li>
<strong>Count matrix</strong>: <code>${sample}_atac_bin.rds</code>
(or <code>.tsv</code>) from <code>get_binned_atac.R</code><br>
</li>
<li>
<strong>Reference</strong>:
<code>Reference/lambdas_ATAC_bincnt.rds</code> (generated via
<code>get_binned_atac.R --generateAggRef</code> or
<code>agg_refs()</code>)<br>
</li>
<li>
<strong>Allele VCF</strong>:
<code>${sample}/${sample}_comb_allele_counts.tsv.gz</code><br>
</li>
<li>
<strong>Genomic bins</strong>: <code>var220kb.rds</code> GRanges
object</li>
</ul>
</li>
<li>
<p><strong>Command</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">Rscript</span> run_numbat_multiome.R <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>  <span class="at">--countmat</span>    <span class="va">${sample}</span>/<span class="va">${sample}</span>_atac_bin.rds <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  <span class="at">--alleledf</span>     <span class="va">${sample}</span>/<span class="va">${sample}</span>_comb_allele_counts.tsv.gz <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="at">--out_dir</span>     <span class="va">${sample}</span>/atac_only/ <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">--ref</span>         Reference/lambdas_ATAC_bincnt.rds <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">--gtf</span>         var220kb.rds <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  <span class="at">--parL</span>        par_numbatm.rds </span></code></pre></div>
<ul>
<li>
<code>--mode ATAC-bin</code> skips RNA likelihoods and relies solely
on ATAC signals.</li>
</ul>
</li>
<li>
<p><strong>Interpretation</strong></p>
<ul>
<li>CNV calls are driven by chromatin-accessibility patterns.<br>
</li>
<li>Ideal when RNA data is unavailable or as a comparison to
Combined-bin.</li>
</ul>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="rna-bin-mode">RNA-bin mode<a class="anchor" aria-label="anchor" href="#rna-bin-mode"></a>
</h3>
<p>In <strong>RNA-bin</strong> mode you leverage only the binned RNA
counts plus the RNA reference:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="va">${numbat_repo}</span>/inst/bin/run_numbat_multiome.R <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  <span class="at">--countmat</span>    <span class="va">${sample}</span>/<span class="va">${sample}</span>_rna_bin.rds <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>  <span class="at">--alleledf</span>     <span class="va">${sample}</span>/<span class="va">${sample}</span>_comb_allele_counts.tsv.gz <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>  <span class="at">--out_dir</span>     <span class="va">${sample}</span>/rna_only/ <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  <span class="at">--ref</span>         Reference/lambdas_RNA_bincnt.rds <span class="dt">\</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="at">--gtf</span>         var220kb.rds <span class="dt">\</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  <span class="at">--parL</span>        par_numbatm.rds </span></code></pre></div>
<ul>
<li>Use when ATAC coverage is low or to benchmark RNA-only CNV
calls.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>For more details, please consult the Numbat-Multiome publication:
&gt; <a href="https://academic.oup.com/bib/article/26/5/bbaf516/8290422" class="external-link">Ruitong
Li, Jean-Baptiste Alberge, Tina Keshavarzian, Junko Tsuji, Johan
Gustafsson, Mahshid Rahmat, Elizabeth D Lightbody, Stephanie L Deng,
Santiago Riviero, Mendy Miller, F Naz Cemre Kalayci, Adrian Wiestner,
Clare Sun, Mathieu Lupien, Irene Ghobrial, Erin Parry, Teng Gao, Gad
Getz. Numbat-multiome: inferring copy number variations by combining RNA
and chromatin accessibility information from single-cell data. Briefings
in Bioinformatics (2025).</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
