<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Interpreting Numbat results • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Interpreting Numbat results">
<meta name="description" content="Visualizing and interpreting Numbat outputs">
<meta property="og:description" content="Visualizing and interpreting Numbat outputs">
<meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/numbat.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a></li>
    <li><a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a></li>
    <li><a class="dropdown-item" href="../articles/descriptions.html">File descriptions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kharchenkolab/numbat/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Interpreting Numbat results</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kharchenkolab/numbat/blob/main/vignettes/results.Rmd" class="external-link"><code>vignettes/results.Rmd</code></a></small>
      <div class="d-none name"><code>results.Rmd</code></div>
    </div>

    
    
<p>In this tutorial, we will illustrate how to visualize and interpret
Numbat output, using a triple-negative breast cancer dataset (TNBC1)
from <a href="https://www.nature.com/articles/s41587-020-00795-2" class="external-link">Gao et
al</a>. We will use <a href="https://github.com/kharchenkolab/pagoda2" class="external-link">pagoda2</a> for
visualizing cells in low-dimensional expression space.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kharchenkolab/numbat/" class="external-link">numbat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com" class="external-link">tidygraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<p>After a Numbat run, we can summarize output files into a Numbat
object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span> <span class="op">=</span> <span class="va"><a href="../reference/Numbat.html">Numbat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>out_dir <span class="op">=</span> <span class="st">'./numbat_out'</span><span class="op">)</span></span></code></pre></div>
<p>In this tutorial, let’s start the analysis with the pre-saved Numbat
and pagoda2 objects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">'http://pklab.org/teng/data/nb_TNBC1.rds'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pagoda</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">'http://pklab.org/teng/data/con_TNBC1.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="copy-number-landscape-and-single-cell-phylogeny">Copy number landscape and single-cell phylogeny<a class="anchor" aria-label="anchor" href="#copy-number-landscape-and-single-cell-phylogeny"></a>
</h2>
<p>As an overview, we can visualize the CNV calls in single-cells and
their evolutionary relationships in an integrated plot panel:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mypal</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1'</span> <span class="op">=</span> <span class="st">'gray'</span>, <span class="st">'2'</span> <span class="op">=</span> <span class="st">"#377EB8"</span>, <span class="st">'3'</span> <span class="op">=</span> <span class="st">"#4DAF4A"</span>, <span class="st">'4'</span> <span class="op">=</span> <span class="st">"#984EA3"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">nb</span><span class="op">$</span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span></span>
<span>  clone_bar <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  p_min <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  pal_clone <span class="op">=</span> <span class="va">mypal</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-5-1.png" width="2400">
In this visualization, the single-cell phylogeny (left) is juxtaposed
with a heatmap of single-cell CNV calls (right). The CNV calls are
colored by the type of alterations (AMP, amplification, BAMP, balanced
amplification, DEL, deletion, CNLoH, copy-neutral loss of
heterozygosity). The colorbar in-between differentiates the distinct
genetic populations (genotype). The dashed blue line separates the
predicted tumor versus normal cells. This tells us that the dataset
mainly consists of three cell populations, a normal population (gray)
and two tumor subclones (green and purple).</p>
<div class="section level3">
<h3 id="refine-subclones-on-the-phylogeny">Refine subclones on the phylogeny<a class="anchor" aria-label="anchor" href="#refine-subclones-on-the-phylogeny"></a>
</h3>
<p>Note that the number of subclones determined by the initial run
parameters in <code>run_numbat</code> can be re-adjusted using
<code>nb$cutree()</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">nb</span><span class="op">$</span><span class="fu">cutree</span><span class="op">(</span>n_cut <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>        <span class="va">nb</span><span class="op">$</span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'n_cut='</span>, <span class="va">k</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-7-1.png" width="2500"></p>
<p>In <code>cutree</code> one can either specify <code>n_cut</code> or
<code>max_cost</code>, which work similarly as <code>k</code> and
<code>h</code> in <code><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">stats::cutree</a></code>. Note that <code>n</code>
cuts should result in <code>n+1</code> clones (the top-level normal
diploid clone is always included). First cut should separate out the
tumor cells as a single clone, second cut gives two tumor subclones, and
so on. Alternatively, one can specify a <code>max_cost</code>, which is
the maximum likelihood cost threshold with which to reduce the phylogeny
(higher <code>max_cost</code> leads to fewer clones).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># restore to original number of cuts</span></span>
<span><span class="va">nb</span><span class="op">$</span><span class="fu">cutree</span><span class="op">(</span>n_cut <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>After refining subclones in the phylogeny, you can recreate the clone
pseduobulk profiles and visualize them (see issue <a href="https://github.com/kharchenkolab/numbat/issues/220" class="external-link">#220</a>).</p>
</div>
</div>
<div class="section level2">
<h2 id="consensus-copy-number-segments">Consensus copy number segments<a class="anchor" aria-label="anchor" href="#consensus-copy-number-segments"></a>
</h2>
<p>Let’s take a look at the consensus segments.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span><span class="op">$</span><span class="fu">plot_consensus</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-9-1.png" width="4200"></p>
</div>
<div class="section level2">
<h2 id="bulk-cnv-profiles">Bulk CNV profiles<a class="anchor" aria-label="anchor" href="#bulk-cnv-profiles"></a>
</h2>
<p>We can also visualize these CNV events in pseudobulks where the data
is more rich, aggregating cells by clone:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span><span class="op">$</span><span class="va">bulk_clones</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n_cells</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_bulks.html">plot_bulks</a></span><span class="op">(</span></span>
<span>    min_LLR <span class="op">=</span> <span class="fl">10</span>, <span class="co"># filtering CNVs by evidence</span></span>
<span>    legend <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-10-1.png" width="3900"></p>
</div>
<div class="section level2">
<h2 id="single-cell-cnv-calls">Single-cell CNV calls<a class="anchor" aria-label="anchor" href="#single-cell-cnv-calls"></a>
</h2>
<p>Numbat probabilistically evaluates the presence/absence of CNVs in
single cells. The cell-level CNV posteriors are stored in the
<code>nb$joint_post</code> dataframe:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nb</span><span class="op">$</span><span class="va">joint_post</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">CHROM</span>, <span class="va">seg</span>, <span class="va">cnv_state</span>, <span class="va">p_cnv</span>, <span class="va">p_cnv_x</span>, <span class="va">p_cnv_y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                        cell  CHROM    seg cnv_state        p_cnv     p_cnv_x</span></span>
<span><span class="co">##                      &lt;char&gt; &lt;fctr&gt; &lt;char&gt;    &lt;char&gt;        &lt;num&gt;       &lt;num&gt;</span></span>
<span><span class="co">## 1: TNBC1_AAACCTGCACCTTGTC-1      1     1b       amp 0.9999999997 1.000000000</span></span>
<span><span class="co">## 2: TNBC1_AAACCTGCACCTTGTC-1      2     2a       amp 0.0002288397 0.001172894</span></span>
<span><span class="co">## 3: TNBC1_AAACCTGCACCTTGTC-1      3     3a       amp 0.0522751360 0.052273344</span></span>
<span><span class="co">## 4: TNBC1_AAACCTGCACCTTGTC-1      3     3c       amp 0.9147696618 0.982278728</span></span>
<span><span class="co">## 5: TNBC1_AAACCTGCACCTTGTC-1      4     4b       amp 0.8999658266 0.845619643</span></span>
<span><span class="co">## 6: TNBC1_AAACCTGCACCTTGTC-1      5     5a       amp 0.8070027725 0.778074328</span></span>
<span><span class="co">##       p_cnv_y</span></span>
<span><span class="co">##         &lt;num&gt;</span></span>
<span><span class="co">## 1: 0.03844235</span></span>
<span><span class="co">## 2: 0.09821435</span></span>
<span><span class="co">## 3: 0.50000000</span></span>
<span><span class="co">## 4: 0.10246769</span></span>
<span><span class="co">## 5: 0.61862288</span></span>
<span><span class="co">## 6: 0.53344303</span></span></code></pre>
<p>which contains cell-level information on specific CNV segments
(<code>seg</code>), their alteration type (<code>cnv_state</code>), the
joint posterior probability of the CNV (<code>p_cnv</code>), the
expression-based posterior (<code>p_cnv_x</code>), and the allele-based
posterior (<code>p_cnv_y</code>). We can visualize the event-specific
posteriors in a expression-based tSNE embedding:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plist</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">muts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1a'</span>, <span class="st">'3b'</span>, <span class="st">'22b'</span><span class="op">)</span></span>
<span><span class="va">cnv_type</span> <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">joint_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">seg</span>, <span class="va">cnv_state</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">cnv_state</span>, <span class="va">.</span><span class="op">$</span><span class="va">seg</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">mut</span> <span class="kw">in</span> <span class="va">muts</span><span class="op">)</span> <span class="op">{</span></span>
<span>    </span>
<span>    <span class="va">plist</span><span class="op">[[</span><span class="va">mut</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">pagoda</span><span class="op">$</span><span class="fu">plotEmbedding</span><span class="op">(</span></span>
<span>        alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span>        size<span class="op">=</span><span class="fl">1</span>, </span>
<span>        plot.na <span class="op">=</span> <span class="cn">F</span>, </span>
<span>        colors <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">joint_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seg</span> <span class="op">==</span> <span class="va">mut</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">p_cnv</span>, <span class="va">.</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">}</span>,</span>
<span>        show.legend <span class="op">=</span> <span class="cn">T</span>,</span>
<span>        mark.groups <span class="op">=</span> <span class="cn">F</span>,</span>
<span>        plot.theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">mut</span>, <span class="st">'('</span>, <span class="va">cnv_type</span><span class="op">[</span><span class="va">muts</span><span class="op">]</span>, <span class="st">')'</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">'royalblue'</span>, mid <span class="op">=</span> <span class="st">'white'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, midpoint <span class="op">=</span> <span class="fl">0.5</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">'Posterior'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plist</span>, guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-12-1.png" width="2700"></p>
</div>
<div class="section level2">
<h2 id="clonal-assignments">Clonal assignments<a class="anchor" aria-label="anchor" href="#clonal-assignments"></a>
</h2>
<p>Numbat aggregates signals across subclone-specific CNVs to
probabilistically assign cells to subclones. The information regarding
clonal assignments are contained in the <code>nb$clone_post</code>
dataframe.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span><span class="op">$</span><span class="va">clone_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">clone_opt</span>, <span class="va">p_1</span>, <span class="va">p_2</span>, <span class="va">p_3</span>, <span class="va">p_4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       cell clone_opt           p_1          p_2          p_3</span></span>
<span><span class="co">## 1 TNBC1_AAACCTGCACCTTGTC-1         4  1.162495e-72 2.105425e-20 3.478936e-29</span></span>
<span><span class="co">## 2 TNBC1_AAACGGGAGTCCTCCT-1         1  1.000000e+00 5.814741e-51 5.920463e-95</span></span>
<span><span class="co">## 3 TNBC1_AAACGGGTCCAGAGGA-1         4 2.434561e-100 2.175022e-14 2.085718e-30</span></span>
<span><span class="co">## 4 TNBC1_AAAGATGCAGTTTACG-1         3  2.020088e-23 8.164920e-04 9.913458e-01</span></span>
<span><span class="co">## 5 TNBC1_AAAGCAACAGGAATGC-1         4  9.546324e-64 4.989037e-11 1.657378e-31</span></span>
<span><span class="co">## 6 TNBC1_AAAGCAATCGGAATCT-1         3  1.516589e-99 8.466806e-33 1.000000e+00</span></span>
<span><span class="co">##            p_4</span></span>
<span><span class="co">## 1 1.000000e+00</span></span>
<span><span class="co">## 2 8.821034e-66</span></span>
<span><span class="co">## 3 1.000000e+00</span></span>
<span><span class="co">## 4 7.837745e-03</span></span>
<span><span class="co">## 5 1.000000e+00</span></span>
<span><span class="co">## 6 2.457592e-29</span></span></code></pre>
<p>Here <code>clone_opt</code> denotes the maximum likelihood assignment
of a cell to a specific clone. <code>p_{1..4}</code> are the detailed
breakdown of the posterior probability that the cell belongs to each
clone, respectively. Let’s visualize the clonal decomposition in a tSNE
embedding. Note that clone 1 is always the normal cells.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pagoda</span><span class="op">$</span><span class="fu">plotEmbedding</span><span class="op">(</span></span>
<span>    alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span>    size<span class="op">=</span><span class="fl">1</span>, </span>
<span>    groups <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">clone_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">clone_opt</span>, <span class="va">.</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">}</span>,</span>
<span>    plot.na <span class="op">=</span> <span class="cn">F</span>,</span>
<span>    plot.theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="st">'Genotypes'</span>,</span>
<span>    pal <span class="op">=</span> <span class="va">mypal</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-14-1.png" width="900"></p>
</div>
<div class="section level2">
<h2 id="tumor-versus-normal-probability">Tumor versus normal probability<a class="anchor" aria-label="anchor" href="#tumor-versus-normal-probability"></a>
</h2>
<p>Combining evidence from all CNVs, Numbat derives an aneuploidy
probability for each cell to distinguish tumor versus normal cells. We
can visualize the posterior aneuploidy probability based on expression
evidence only, allele evidence only, and jointly:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_joint</span> <span class="op">=</span> <span class="va">pagoda</span><span class="op">$</span><span class="fu">plotEmbedding</span><span class="op">(</span></span>
<span>    alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span>    size<span class="op">=</span><span class="fl">1</span>, </span>
<span>    colors <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">clone_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">p_cnv</span>, <span class="va">.</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">}</span>,</span>
<span>    plot.na <span class="op">=</span> <span class="cn">F</span>,</span>
<span>    plot.theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="st">'Joint'</span>,</span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">'royalblue'</span>, mid <span class="op">=</span> <span class="st">'white'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, midpoint <span class="op">=</span> <span class="fl">0.5</span>, name <span class="op">=</span> <span class="st">'Posterior'</span><span class="op">)</span></span>
<span><span class="va">p_allele</span> <span class="op">=</span> <span class="va">pagoda</span><span class="op">$</span><span class="fu">plotEmbedding</span><span class="op">(</span></span>
<span>    alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span>    size<span class="op">=</span><span class="fl">1</span>, </span>
<span>    colors <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">clone_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">p_cnv_x</span>, <span class="va">.</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">}</span>,</span>
<span>    plot.na <span class="op">=</span> <span class="cn">F</span>,</span>
<span>    plot.theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="st">'Expression'</span>,</span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">'royalblue'</span>, mid <span class="op">=</span> <span class="st">'white'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, midpoint <span class="op">=</span> <span class="fl">0.5</span>, name <span class="op">=</span> <span class="st">'Posterior'</span><span class="op">)</span></span>
<span><span class="va">p_expr</span> <span class="op">=</span> <span class="va">pagoda</span><span class="op">$</span><span class="fu">plotEmbedding</span><span class="op">(</span></span>
<span>    alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span>    size<span class="op">=</span><span class="fl">1</span>, </span>
<span>    colors <span class="op">=</span> <span class="va">nb</span><span class="op">$</span><span class="va">clone_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">p_cnv_y</span>, <span class="va">.</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">}</span>,</span>
<span>    plot.na <span class="op">=</span> <span class="cn">F</span>,</span>
<span>    show.legend <span class="op">=</span> <span class="cn">T</span>,</span>
<span>    plot.theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="st">'Allele'</span>,</span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">'royalblue'</span>, mid <span class="op">=</span> <span class="st">'white'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, midpoint <span class="op">=</span> <span class="fl">0.5</span>, name <span class="op">=</span> <span class="st">'Posterior'</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">p_expr</span> <span class="op">|</span> <span class="va">p_allele</span> <span class="op">|</span> <span class="va">p_joint</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-15-1.png" width="2700">
Both expression and allele signal clearly separate the tumor and normal
cells.</p>
</div>
<div class="section level2">
<h2 id="tumor-phylogeny">Tumor phylogeny<a class="anchor" aria-label="anchor" href="#tumor-phylogeny"></a>
</h2>
<p>Let’s take a closer look at the inferred single cell phylogeny and
where the mutations occurred.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span><span class="op">$</span><span class="fu">plot_sc_tree</span><span class="op">(</span></span>
<span>  label_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  branch_width <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  tip_length <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  pal_clone <span class="op">=</span> <span class="va">mypal</span>, </span>
<span>  tip <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-16-1.png" width="1800">
The mutational history can also be represented on the clone level, where
cells with the same genotype are aggregated into one node.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span><span class="op">$</span><span class="fu">plot_mut_history</span><span class="op">(</span>pal <span class="op">=</span> <span class="va">mypal</span><span class="op">)</span></span></code></pre></div>
<p><img src="results_files/figure-html/unnamed-chunk-17-1.png" width="1800"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
