<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Allele-specific CNV analysis from spatial transcriptomics">
<title>Spatial transcriptomics • numbat</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial transcriptomics">
<meta property="og:description" content="Allele-specific CNV analysis from spatial transcriptomics">
<meta property="og:image" content="https://kharchenkolab.github.io/numbat/logo.png">
<meta name="google-site-verification" content="GE78cdBqx5YPEbHjqtohe_p51Ng8oEPXdcaSEIu2dP8">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">numbat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/numbat.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/results.html">Interpreting Numbat results</a>
    <a class="dropdown-item" href="../articles/mouse.html">Mouse tutorial</a>
    <a class="dropdown-item" href="../articles/spatial-rna.html">Spatial transcriptomics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Spatial transcriptomics</h1>
            
      
      
      <div class="d-none name"><code>spatial-rna.Rmd</code></div>
    </div>

    
    
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:site" content="@tGaoTeng">
<meta name="twitter:title" content="Numbat spatial">
<meta name="twitter:description" content="Allele-specific CNV analysis from spatial transcriptomics">
<meta name="twitter:image" content="https://i.imgur.com/6n2m6AS.png">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Spatial transcriptomics (ST) assays provide an ideal opportunity to
characterize the spatial architecture of a tumor as well as the
cell-cell interactions within the tumor microenvironment. From
transcriptomic information, CNV analysis can be applied to resolve
genetic subclones in a spatial context <a href="#references">[1,2]</a>.
In particular, many sequencing-based ST assays (e.g. Visium for fresh
frozen and Slide-seq) capture both expression and allele information,
and are thus amenable to allele-specific CNV analysis.</p>
<p>Unlike single-cell sequencing, measurements from spatial assays are
often from a mixture cells (i.e. “spots”). Such mixtures can potentially
consist of tumor and normal cells (stromal/immune) as well as different
tumor subclones with distinct genotypes. Originally developed for
scRNA-seq data, Numbat makes several assumptions in the model based on
the single-cell nature of the measurements. However, if we are willing
to accept the assumption that most spots in ST only contain cells from
the same clone (genetic homogeneity within a spot), then Numbat is
applicable to ST data. Whether this assumption is reasonable likely
depends on the technology (for example, Visium spots typically contain
1-10 cells, whereas Slide-seqV2 is near single-cell resolution) as well
as tissue architecture (some tissues are more well-mixed than others).
It is worth noting that existing spatial CNV calling methods such as
STARCH <a href="#references">[3]</a> also make the same assumption.
Furthermore, future generations of ST assays will likely achieve
single-cell resolution very soon.</p>
<p>As 10x Visium is one of the most widely used ST platforms, this
vignette is dedicated to the application of Numbat to perform
allele-specific CNV analysis on Visium data. Recently published data
from Barkely et al <a href="#references">[1]</a> (downloadable at <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE203612" class="external-link">GSE203612</a>)
will be used for illustration.</p>
</div>
<div class="section level2">
<h2 id="preparing-data">Preparing data<a class="anchor" aria-label="anchor" href="#preparing-data"></a>
</h2>
<p>Because outputs from 10x SpaceRanger (BAM, barcode list, and gene
expression quantification) are in the same format as CellRanger, the
same <a href="https://kharchenkolab.github.io/numbat/articles/numbat.html#preparing-data"><code>pileup-and-phase</code></a>
procedure can be used to prepare the allele counts and expression count
matrices for Visium data as scRNA-seq. Compared to scRNA-seq, the
overall SNP coverage from Visium is typically more sparse, which may
impact phasing quality. Therefore, a priori genotyping by DNA or
co-genotyping with scRNA-seq (via multi-sample mode of
<code>pileup-and-phase</code>) would be especially useful.</p>
<p>In terms of expression references, reasonable choices include Visium
or scRNA-seq data of matching normal tissues (e.g. normal prostate as
reference for prostate cancer samples). For Visium data, one can first
cluster the spots (e.g. using <code>pagoda2</code>) and then collapse
the counts by clusters (using <a href="https://kharchenkolab.github.io/numbat/reference/aggregate_counts.html"><code>numbat::aggregate_counts</code></a>).
For Visium, other Visium data usually make a better reference than
scRNA-seq data, achieving the least amount of noise. For this anlaysis,
I just used clustered spots from UCEC3 as reference, since that sample
has no aneuploidy.</p>
</div>
<div class="section level2">
<h2 id="running-analysis">Running analysis<a class="anchor" aria-label="anchor" href="#running-analysis"></a>
</h2>
<p>The recommended parameters for Visium are similar to those for
scRNA-seq. The only difference is that, because allele coverage per
barcode is typically sparser than scRNA-seq, <code>max_entropy</code>
(which quantifies the uncertainty in CNV posterior in individual
cells/spots) should be set to a higher value such as 0.8 (default is
0.5).</p>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<div class="section level3">
<h3 id="aneuploidy-probabilities">Aneuploidy probabilities<a class="anchor" aria-label="anchor" href="#aneuploidy-probabilities"></a>
</h3>
<p>A major task in ST analysis is delineating the boundaries of
malignant and normal cells. Aggregating all CNV evidence, Numbat outputs
a mutant versus normal <em>probability</em>
(<code>clone_post$p_cnv</code>). This provides a natural interpretation
for spots that are a mixture of mutant and normal cells, which should
show a CNV probability around 0.5 (the specific value depends on the
mixture fraction). Spots that contain purely mutant cells should show a
high CNV probability (near 1) and those that contain purely normal cells
should show a low CNV probability (near 0).</p>
<p>Let’s first look at the mutant versus normal probabilities of a few
samples.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidyverse/glue" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kharchenkolab.github.io/numbat">numbat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read in the numbat results and spot coordinates</span></span>
<span><span class="va">nb</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">spots</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'OVCA1'</span>, <span class="st">'OVCA3'</span>, <span class="st">'BRCA1'</span>, <span class="st">'GIST1'</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">sample</span> <span class="kw">in</span> <span class="va">samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">nb</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va"><a href="../reference/Numbat.html">Numbat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'~/results/{sample}'</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">spots</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">'~/{sample}/outs/spatial/tissue_positions.csv'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">8.5</span>, repr.plot.height <span class="op">=</span> <span class="fl">8</span>, repr.plot.res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'OVCA1'</span>, <span class="st">'OVCA3'</span>, <span class="st">'GIST1'</span>, <span class="st">'BRCA1'</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">{</span></span>
<span>        </span>
<span>        <span class="va">spots</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>                <span class="va">nb</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clone_post</span>, </span>
<span>                by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span> <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span>            <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_row</span>, y <span class="op">=</span> <span class="va">array_col</span><span class="op">)</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">p_cnv</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span></span>
<span>                low <span class="op">=</span> <span class="st">'darkgreen'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, mid <span class="op">=</span> <span class="st">'yellow'</span>, </span>
<span>                midpoint <span class="op">=</span> <span class="fl">0.5</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">oob_squish</a></span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">77</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">127</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/hZpNWVJ.jpg"></p>
<p><strong>Figure 1. Mutant versus normal probabilities.</strong> Red
represents mutant (aneuploid), green represents normal, and yellow
represents mixture/uncertain.</p>
<p>Overlaying with the tissue images, we can see that the tumor/normal
boundaries coincide with visible tissue architecture.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">8.5</span>, repr.plot.height <span class="op">=</span> <span class="fl">8</span>, repr.plot.res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'OVCA1'</span>, <span class="st">'OVCA3'</span>, <span class="st">'GIST1'</span>, <span class="st">'BRCA1'</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">{</span></span>
<span>        </span>
<span>        <span class="va">img</span> <span class="op">&lt;-</span> <span class="fu">jpeg</span><span class="fu">::</span><span class="fu">readJPEG</span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"~/visium/nyu/images/{sample}_cropped.jpg"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="va">spots</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>                <span class="va">nb</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clone_post</span>, </span>
<span>                by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span> <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span>            <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_row</span>, y <span class="op">=</span> <span class="va">array_col</span><span class="op">)</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">background_image</span><span class="op">(</span><span class="va">img</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">p_cnv</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                stroke <span class="op">=</span> <span class="fl">0.3</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, pch <span class="op">=</span> <span class="fl">21</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span></span>
<span>                low <span class="op">=</span> <span class="st">'darkgreen'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, mid <span class="op">=</span> <span class="st">'yellow'</span>, </span>
<span>                midpoint <span class="op">=</span> <span class="fl">0.5</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">oob_squish</a></span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">77</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">127</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/NsBjisZ.jpg"><strong>Figure 2.
Mutant versus normal probabilities overlayed with tissue
images.</strong> Red represents mutant (aneuploid), green represents
normal, and yellow represents mixture/uncertain.</p>
</div>
<div class="section level3">
<h3 id="per-event-cnv-probabilities">Per-event CNV probabilities<a class="anchor" aria-label="anchor" href="#per-event-cnv-probabilities"></a>
</h3>
<p>In addition to overall aneuploidy probability, Numbat also outputs a
posterior probability per CNV event for each spot
(<code>joint_post$p_cnv</code>). This is especially useful for resolving
tumor subclonal structures. Let’s take BRCA0 as an example and plot the
per-event CNV probabilities:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">11</span>, repr.plot.height <span class="op">=</span> <span class="fl">8</span>, repr.plot.res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample</span> <span class="op">=</span> <span class="st">'BRCA0'</span></span>
<span></span>
<span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu">jpeg</span><span class="fu">::</span><span class="fu">readJPEG</span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"~/visium/nyu/images/{sample}_cropped.jpg"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spots</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>        <span class="va">nb</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">joint_post</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">LLR</span> <span class="op">&gt;</span> <span class="fl">35</span> <span class="op">&amp;</span> <span class="va">avg_entropy</span> <span class="op">&lt;</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>p_total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">p_cnv</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">p_total</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>seg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">seg</span>, <span class="st">'('</span>, <span class="va">cnv_state</span>, <span class="st">')'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>seg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seg</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, </span>
<span>        by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span> <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_row</span>, y <span class="op">=</span> <span class="va">array_col</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">p_cnv</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.1</span>, stroke <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span></span>
<span>        low <span class="op">=</span> <span class="st">'blue'</span>, high <span class="op">=</span> <span class="st">'red3'</span>, mid <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>        midpoint <span class="op">=</span> <span class="fl">0.5</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">oob_squish</a></span></span>
<span>    <span class="op">)</span> <span class="op">+</span>   </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">seg</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">77</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">127</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/aAylHMO.jpg"><strong>Figure 3.
Probabilities of specific CNV events.</strong> Red represents high
probability of event presence, blue represents low probability of event
presence, and white represents mixture/uncertain.</p>
<p>We see that there is substantial heterogeneity among CNV events in
their spatial distribution. In particular, deletions on chromosomes 9,
14, 16, 17, 19 and 22 seem to mark specific patches of the tissue
section.</p>
</div>
<div class="section level3">
<h3 id="tumor-subclones">Tumor subclones<a class="anchor" aria-label="anchor" href="#tumor-subclones"></a>
</h3>
<p>Finally, we can visualize the tumor subclonal structure inferred by
Numbat in spatial coordinates. First, let’s look at the usual
phylo-heatmap which provides a summary of the clonal structure.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">8</span>, repr.plot.height <span class="op">=</span> <span class="fl">4</span>, repr.plot.res <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pal</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="st">"gray"</span>, `2` <span class="op">=</span> <span class="st">"#9E0142"</span>, `3` <span class="op">=</span> <span class="st">"#F67D4A"</span>,</span>
<span>        `4` <span class="op">=</span> <span class="st">"#F2EA91"</span>, `5` <span class="op">=</span> <span class="st">"#77C8A4"</span>, `6` <span class="op">=</span> <span class="st">"#5E4FA2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nb</span><span class="op">[[</span><span class="st">'BRCA0'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="fu">plot_phylo_heatmap</span><span class="op">(</span>pal_clone <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/wVjnjpf.png"><strong>Figure 4. Tumor
phylogeny and CNV heatmap of BRCA0.</strong> The stacked color bars on
the left reflect clonal assignment probabilities.</p>
<p>From the phylo-heatmap, we can see that clone 5 in particular harbors
significantly more CNV events than the rest of the tumor cells. Let’s
visualize the spatial distribution of subclones (clone_post$clone_opt)
and overlay with tissue image.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.width <span class="op">=</span> <span class="fl">5</span>, repr.plot.height <span class="op">=</span> <span class="fl">4.2</span>, repr.plot.res <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample</span> <span class="op">=</span> <span class="st">'BRCA0'</span></span>
<span></span>
<span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu">jpeg</span><span class="fu">::</span><span class="fu">readJPEG</span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"~/visium/nyu/images/{sample}_cropped.jpg"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spots</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>        <span class="va">nb</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clone_post</span>, </span>
<span>        by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span> <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>clone_opt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clone_opt</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_row</span>, y <span class="op">=</span> <span class="va">array_col</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">background_image</span><span class="op">(</span><span class="va">img</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">clone_opt</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.3</span>, stroke <span class="op">=</span> <span class="fl">0.3</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, pch <span class="op">=</span> <span class="fl">21</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">pal</span>, limits <span class="op">=</span> <span class="va">force</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span></span>
<span>        color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clone"</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">77</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">127</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
<p><img src="https://i.imgur.com/shtqgqo.jpg"><strong>Figure 5.
Clonal assignments in BRCA0.</strong> The maximum a posteriori clonal
assignments of individual spots are shown.</p>
<p>Interestingly, if you zoom in close enough, you can see that the
distribution of clone 5 corresponds to visible foci that are clearly
distinct from the surrounding tumor cells. Taken together, these
observations suggest that clone 5 represents a further evolved tumor
subpopulation that is actively invading the diseased tissue.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Like for scRNA-seq, allele-specific CNV analysis can be applied to
spatial transcriptomics to interrogate genetic heterogeneity in a tumor.
We have shown that Numbat can be used to differentiate (aneuploid) tumor
and stromal compartments, evaluate the presence/absence of specific CNV
events in spatial locations, and elucidate the spatial distributions of
tumor subclones. Because Visium data is even sparser than scRNA-seq, the
added information provided by phased allele counts is especially
valuable for CNV calling. The key assumption we have to keep in mind is
that cells contained in a spot are assumed to be largely genetically
homogenous. However, this assumption is not that limiting in practice
because cell mixtures are reflected as uncertainty in the posterior
probability of CNV presence as well as clonal assignments. Newer
generations of ST technologies that provide single-cell resolution can
help overcome this limitation in the future.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>[1] Barkley, D., Moncada, R., Pour, M. et al. Cancer cell states
recur across tumor types and form specific interactions with the tumor
microenvironment. Nat Genet 54, 1192–1201 (2022). <a href="https://doi.org/10.1038/s41588-022-01141-9" class="external-link uri">https://doi.org/10.1038/s41588-022-01141-9</a></p>
<p>[2] Wei, R., He, S., Bai, S. et al. Spatial charting of single-cell
transcriptomes in tissues. Nat Biotechnol 40, 1190–1199 (2022). <a href="https://www.nature.com/articles/s41587-022-01233-1" class="external-link uri">https://www.nature.com/articles/s41587-022-01233-1</a></p>
<p>[3] Elyanow R, Zeira R, Land M, Raphael BJ. STARCH: copy number and
clone inference from spatial transcriptomics data. Phys Biol. 2021 Mar
9;18(3):035001. doi: 10.1088/1478-3975/abbe99. PMID: 33022659.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Teng Gao, Ruslan Soldatov, Hirak Sarkar, Evan Biederstedt, Peter Kharchenko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
